{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a58d01f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "\n",
    "from os import path\n",
    "from pathlib import Path  \n",
    "\n",
    "#pd.set_option('display.max_columns', None)\n",
    "#pd.set_option('display.max_rows', None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "586180b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import files\n",
    "data_file = path.join('/Users/louxuwen/Desktop/Documents/GitHub/SP23-BENG213/Project/Map Gene to KEGG Pathway/down_data_matrix.csv') # Data matrix\n",
    "gene_file = path.join('/Users/louxuwen/Desktop/Documents/GitHub/SP23-BENG213/Project/Map Gene to KEGG Pathway/Homo_sapiens.tsv') # Gene Symbol and ID\n",
    "pathway_file = path.join('/Users/louxuwen/Desktop/Documents/GitHub/SP23-BENG213/Project/Map Gene to KEGG Pathway/Pathway_Down.csv') # all the pathways used in the heatmap\n",
    "gene_to_pathway_file = path.join('/Users/louxuwen/Desktop/Documents/GitHub/SP23-BENG213/Project/Map Gene to KEGG Pathway/Gene_to_Pathway.csv') # every pathway and the relevant genes\n",
    "\n",
    "# dataframe of metadata\n",
    "data = pd.read_csv(data_file,index_col=0)\n",
    "gene_info = pd.read_csv(gene_file, sep=\"\\t\")\n",
    "pathway = pd.read_csv(pathway_file,index_col=0).fillna(\"0\")\n",
    "gene_to_pathway = pd.read_csv(gene_to_pathway_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "9f914b3b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Pathway</th>\n",
       "      <th>Gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>hsa00010</td>\n",
       "      <td>10327</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>hsa00010</td>\n",
       "      <td>124</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>hsa00010</td>\n",
       "      <td>125</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>hsa00010</td>\n",
       "      <td>126</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>hsa00010</td>\n",
       "      <td>127</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36117</th>\n",
       "      <td>hsa05418</td>\n",
       "      <td>91860</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36118</th>\n",
       "      <td>hsa05418</td>\n",
       "      <td>92</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36119</th>\n",
       "      <td>hsa05418</td>\n",
       "      <td>93</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36120</th>\n",
       "      <td>hsa05418</td>\n",
       "      <td>9446</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36121</th>\n",
       "      <td>hsa05418</td>\n",
       "      <td>9817</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>36122 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Pathway   Gene\n",
       "0      hsa00010  10327\n",
       "1      hsa00010    124\n",
       "2      hsa00010    125\n",
       "3      hsa00010    126\n",
       "4      hsa00010    127\n",
       "...         ...    ...\n",
       "36117  hsa05418  91860\n",
       "36118  hsa05418     92\n",
       "36119  hsa05418     93\n",
       "36120  hsa05418   9446\n",
       "36121  hsa05418   9817\n",
       "\n",
       "[36122 rows x 2 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_to_pathway"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "3de20c86",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#tax_id</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>Symbol</th>\n",
       "      <th>LocusTag</th>\n",
       "      <th>Synonyms</th>\n",
       "      <th>dbXrefs</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>map_location</th>\n",
       "      <th>description</th>\n",
       "      <th>type_of_gene</th>\n",
       "      <th>Symbol_from_nomenclature_authority</th>\n",
       "      <th>Full_name_from_nomenclature_authority</th>\n",
       "      <th>Nomenclature_status</th>\n",
       "      <th>Other_designations</th>\n",
       "      <th>Modification_date</th>\n",
       "      <th>Feature_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9606</td>\n",
       "      <td>1</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>-</td>\n",
       "      <td>A1B|ABG|GAB|HYST2477</td>\n",
       "      <td>MIM:138670|HGNC:HGNC:5|Ensembl:ENSG00000121410...</td>\n",
       "      <td>19</td>\n",
       "      <td>19q13.43</td>\n",
       "      <td>alpha-1-B glycoprotein</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>alpha-1-B glycoprotein</td>\n",
       "      <td>O</td>\n",
       "      <td>alpha-1B-glycoprotein|HEL-S-163pA|epididymis s...</td>\n",
       "      <td>20230329</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>9606</td>\n",
       "      <td>2</td>\n",
       "      <td>A2M</td>\n",
       "      <td>-</td>\n",
       "      <td>A2MD|CPAMD5|FWP007|S863-7</td>\n",
       "      <td>MIM:103950|HGNC:HGNC:7|Ensembl:ENSG00000175899...</td>\n",
       "      <td>12</td>\n",
       "      <td>12p13.31</td>\n",
       "      <td>alpha-2-macroglobulin</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>A2M</td>\n",
       "      <td>alpha-2-macroglobulin</td>\n",
       "      <td>O</td>\n",
       "      <td>alpha-2-macroglobulin|C3 and PZP-like alpha-2-...</td>\n",
       "      <td>20230405</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>9606</td>\n",
       "      <td>3</td>\n",
       "      <td>A2MP1</td>\n",
       "      <td>-</td>\n",
       "      <td>A2MP</td>\n",
       "      <td>HGNC:HGNC:8|Ensembl:ENSG00000291190|AllianceGe...</td>\n",
       "      <td>12</td>\n",
       "      <td>12p13.31</td>\n",
       "      <td>alpha-2-macroglobulin pseudogene 1</td>\n",
       "      <td>pseudo</td>\n",
       "      <td>A2MP1</td>\n",
       "      <td>alpha-2-macroglobulin pseudogene 1</td>\n",
       "      <td>O</td>\n",
       "      <td>pregnancy-zone protein pseudogene</td>\n",
       "      <td>20230329</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>9606</td>\n",
       "      <td>9</td>\n",
       "      <td>NAT1</td>\n",
       "      <td>-</td>\n",
       "      <td>AAC1|MNAT|NAT-1|NATI</td>\n",
       "      <td>MIM:108345|HGNC:HGNC:7645|Ensembl:ENSG00000171...</td>\n",
       "      <td>8</td>\n",
       "      <td>8p22</td>\n",
       "      <td>N-acetyltransferase 1</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NAT1</td>\n",
       "      <td>N-acetyltransferase 1</td>\n",
       "      <td>O</td>\n",
       "      <td>arylamine N-acetyltransferase 1|N-acetyltransf...</td>\n",
       "      <td>20230509</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>9606</td>\n",
       "      <td>10</td>\n",
       "      <td>NAT2</td>\n",
       "      <td>-</td>\n",
       "      <td>AAC2|NAT-2|PNAT</td>\n",
       "      <td>MIM:612182|HGNC:HGNC:7646|Ensembl:ENSG00000156...</td>\n",
       "      <td>8</td>\n",
       "      <td>8p22</td>\n",
       "      <td>N-acetyltransferase 2</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NAT2</td>\n",
       "      <td>N-acetyltransferase 2</td>\n",
       "      <td>O</td>\n",
       "      <td>arylamine N-acetyltransferase 2|N-acetyltransf...</td>\n",
       "      <td>20230509</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>169029</th>\n",
       "      <td>741158</td>\n",
       "      <td>8923215</td>\n",
       "      <td>trnD</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>MT</td>\n",
       "      <td>-</td>\n",
       "      <td>tRNA-Asp</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>20200909</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>169030</th>\n",
       "      <td>741158</td>\n",
       "      <td>8923216</td>\n",
       "      <td>trnP</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>MT</td>\n",
       "      <td>-</td>\n",
       "      <td>tRNA-Pro</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>20200909</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>169031</th>\n",
       "      <td>741158</td>\n",
       "      <td>8923217</td>\n",
       "      <td>trnA</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>MT</td>\n",
       "      <td>-</td>\n",
       "      <td>tRNA-Ala</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>20200909</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>169032</th>\n",
       "      <td>741158</td>\n",
       "      <td>8923218</td>\n",
       "      <td>COX1</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>MT</td>\n",
       "      <td>-</td>\n",
       "      <td>cytochrome c oxidase subunit I</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>cytochrome c oxidase subunit I</td>\n",
       "      <td>20200909</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>169033</th>\n",
       "      <td>741158</td>\n",
       "      <td>8923219</td>\n",
       "      <td>16S rRNA</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>MT</td>\n",
       "      <td>-</td>\n",
       "      <td>l-rRNA</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>20200909</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>169034 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        #tax_id   GeneID    Symbol LocusTag                   Synonyms  \\\n",
       "0          9606        1      A1BG        -       A1B|ABG|GAB|HYST2477   \n",
       "1          9606        2       A2M        -  A2MD|CPAMD5|FWP007|S863-7   \n",
       "2          9606        3     A2MP1        -                       A2MP   \n",
       "3          9606        9      NAT1        -       AAC1|MNAT|NAT-1|NATI   \n",
       "4          9606       10      NAT2        -            AAC2|NAT-2|PNAT   \n",
       "...         ...      ...       ...      ...                        ...   \n",
       "169029   741158  8923215      trnD        -                          -   \n",
       "169030   741158  8923216      trnP        -                          -   \n",
       "169031   741158  8923217      trnA        -                          -   \n",
       "169032   741158  8923218      COX1        -                          -   \n",
       "169033   741158  8923219  16S rRNA        -                          -   \n",
       "\n",
       "                                                  dbXrefs chromosome  \\\n",
       "0       MIM:138670|HGNC:HGNC:5|Ensembl:ENSG00000121410...         19   \n",
       "1       MIM:103950|HGNC:HGNC:7|Ensembl:ENSG00000175899...         12   \n",
       "2       HGNC:HGNC:8|Ensembl:ENSG00000291190|AllianceGe...         12   \n",
       "3       MIM:108345|HGNC:HGNC:7645|Ensembl:ENSG00000171...          8   \n",
       "4       MIM:612182|HGNC:HGNC:7646|Ensembl:ENSG00000156...          8   \n",
       "...                                                   ...        ...   \n",
       "169029                                                  -         MT   \n",
       "169030                                                  -         MT   \n",
       "169031                                                  -         MT   \n",
       "169032                                                  -         MT   \n",
       "169033                                                  -         MT   \n",
       "\n",
       "       map_location                         description    type_of_gene  \\\n",
       "0          19q13.43              alpha-1-B glycoprotein  protein-coding   \n",
       "1          12p13.31               alpha-2-macroglobulin  protein-coding   \n",
       "2          12p13.31  alpha-2-macroglobulin pseudogene 1          pseudo   \n",
       "3              8p22               N-acetyltransferase 1  protein-coding   \n",
       "4              8p22               N-acetyltransferase 2  protein-coding   \n",
       "...             ...                                 ...             ...   \n",
       "169029            -                            tRNA-Asp            tRNA   \n",
       "169030            -                            tRNA-Pro            tRNA   \n",
       "169031            -                            tRNA-Ala            tRNA   \n",
       "169032            -      cytochrome c oxidase subunit I  protein-coding   \n",
       "169033            -                              l-rRNA            rRNA   \n",
       "\n",
       "       Symbol_from_nomenclature_authority  \\\n",
       "0                                    A1BG   \n",
       "1                                     A2M   \n",
       "2                                   A2MP1   \n",
       "3                                    NAT1   \n",
       "4                                    NAT2   \n",
       "...                                   ...   \n",
       "169029                                  -   \n",
       "169030                                  -   \n",
       "169031                                  -   \n",
       "169032                                  -   \n",
       "169033                                  -   \n",
       "\n",
       "       Full_name_from_nomenclature_authority Nomenclature_status  \\\n",
       "0                     alpha-1-B glycoprotein                   O   \n",
       "1                      alpha-2-macroglobulin                   O   \n",
       "2         alpha-2-macroglobulin pseudogene 1                   O   \n",
       "3                      N-acetyltransferase 1                   O   \n",
       "4                      N-acetyltransferase 2                   O   \n",
       "...                                      ...                 ...   \n",
       "169029                                     -                   -   \n",
       "169030                                     -                   -   \n",
       "169031                                     -                   -   \n",
       "169032                                     -                   -   \n",
       "169033                                     -                   -   \n",
       "\n",
       "                                       Other_designations  Modification_date  \\\n",
       "0       alpha-1B-glycoprotein|HEL-S-163pA|epididymis s...           20230329   \n",
       "1       alpha-2-macroglobulin|C3 and PZP-like alpha-2-...           20230405   \n",
       "2                       pregnancy-zone protein pseudogene           20230329   \n",
       "3       arylamine N-acetyltransferase 1|N-acetyltransf...           20230509   \n",
       "4       arylamine N-acetyltransferase 2|N-acetyltransf...           20230509   \n",
       "...                                                   ...                ...   \n",
       "169029                                                  -           20200909   \n",
       "169030                                                  -           20200909   \n",
       "169031                                                  -           20200909   \n",
       "169032                     cytochrome c oxidase subunit I           20200909   \n",
       "169033                                                  -           20200909   \n",
       "\n",
       "       Feature_type  \n",
       "0                 -  \n",
       "1                 -  \n",
       "2                 -  \n",
       "3                 -  \n",
       "4                 -  \n",
       "...             ...  \n",
       "169029            -  \n",
       "169030            -  \n",
       "169031            -  \n",
       "169032            -  \n",
       "169033            -  \n",
       "\n",
       "[169034 rows x 16 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ebe75ab4",
   "metadata": {},
   "outputs": [],
   "source": [
    "pathways = pathway.index.tolist()\n",
    "pathway_name = []\n",
    "pathway_number = []\n",
    "for i in range(len(pathways)):\n",
    "    temp = pathway.iloc[i]\n",
    "    temp2 = []\n",
    "    for j in range(4):\n",
    "        if temp[j] != \"0\":\n",
    "            pathway_name.append(pathways[i])\n",
    "            pathway_number.append(temp[j])\n",
    "pathway_number = set(pathway_number)            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dddbd0f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get gene IDs\n",
    "# note that 7 percent (~252/3418 genes does not have matching symbol=ID)\n",
    "Gene_IDs = []\n",
    "for i in range(len(data.index.tolist())):\n",
    "    Gene_name = data.iloc[i][0]\n",
    "    \n",
    "    # get gene symbol\n",
    "    if \",\" in Gene_name:\n",
    "        temp = Gene_name.split(\",\")\n",
    "        if temp[0] != \"\":\n",
    "            Gene_name = temp[0]\n",
    "        else:\n",
    "            Gene_name = temp[1].lstrip()\n",
    "    \n",
    "    # gene_symbol to ID\n",
    "    ID = gene_info.loc[gene_info['Symbol'] == Gene_name]['GeneID'].tolist()\n",
    "    if ID == []:\n",
    "        Gene_IDs.append(\"None\")\n",
    "    else:\n",
    "        Gene_IDs.append(ID[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f7f5022d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# add the IDs column in and drop the genes that doesn't have an ID\n",
    "data[\"ID\"] = Gene_IDs\n",
    "data = data[data['ID'] != \"None\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a2fc2685",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get the pathway\n",
    "# only 1402/3166 genes belong to a functional pathway\n",
    "pathways_by_gene = []\n",
    "pathways_by_gene_column = []\n",
    "for i in range(len(data.index.tolist())):\n",
    "    temp_ID = data.iloc[i][28]\n",
    "    \n",
    "    # get corresponding pathways\n",
    "    temp_pathways = gene_to_pathway[gene_to_pathway['Gene']== data.iloc[i][28]][\"Pathway\"].tolist()\n",
    "    \n",
    "    if len(temp_pathways) == 0:\n",
    "        pathways_by_gene_column.append(\"none\")\n",
    "    else:\n",
    "        pathways_by_gene.append(temp_pathways)\n",
    "        pathways_by_gene_column.append(\" \".join(temp_pathways))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "98f4b5fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "data[\"All Pathways\"] = pathways_by_gene_column\n",
    "data = data[data['All Pathways'] != \"none\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0f97b023",
   "metadata": {},
   "source": [
    "## output the data file as long as the gene is involved with a pathways ##"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6b32f7ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "filepath = Path('/Users/louxuwen/Desktop/Documents/GitHub/SP23-BENG213/Project/Map Gene to KEGG Pathway/Data_All_Pathways.csv')  \n",
    "filepath.parent.mkdir(parents=True, exist_ok=True)  \n",
    "data.to_csv(filepath)  "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4b0192cb",
   "metadata": {},
   "source": [
    "## continue to shrink data only if a gene is involved with a pathway actually used in the heatmap ##"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b68696ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "# find the pathway that's involved with the heat map\n",
    "pathways_involved = []\n",
    "pathways_involved_column = []\n",
    "for i in range(len(data.index.tolist())):\n",
    "    a = pathways_by_gene[i]\n",
    "    temp = []\n",
    "\n",
    "    for j in a:\n",
    "        if j in pathway_number:\n",
    "            temp.append(j)\n",
    "    \n",
    "    if len(temp) == 0:\n",
    "        pathways_involved_column.append(\"none\")\n",
    "    else:\n",
    "        pathways_involved.append(temp)\n",
    "        pathways_involved_column.append(\" \".join(temp))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "334aea36",
   "metadata": {},
   "outputs": [],
   "source": [
    "data[\"Involved Pathways\"] = pathways_involved_column\n",
    "data = data[data['Involved Pathways'] != \"none\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6a1ccf1f",
   "metadata": {},
   "outputs": [],
   "source": [
    "drugs = data.columns.values.tolist()[1:28]\n",
    "functions = pathway.index.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bf0ca0b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create the finalmatrix\n",
    "final_matrix = [[0.0] * len(drugs)]*len(functions)\n",
    "counter=0\n",
    "for i in range(len(data.index.tolist())):\n",
    "    temp_pathways = pathways_involved[i]\n",
    "    temp_values = data.iloc[i].tolist()[1:28]\n",
    "    temp_functions = []\n",
    "    for j in temp_pathways:\n",
    "        a = pathway[pathway[\"pathway_1\"] == j].index.tolist()\n",
    "        if len(a)>0:\n",
    "            temp_functions.append(a[0])\n",
    "        else:\n",
    "            a = pathway[pathway[\"pathway_2\"] == j].index.tolist()\n",
    "            if len(a)>0:\n",
    "                temp_functions.append(a[0])\n",
    "            else:\n",
    "                a = pathway[pathway[\"pathway_3\"] == j].index.tolist()\n",
    "                if len(a)>0:\n",
    "                    temp_functions.append(a[0])\n",
    "                else:\n",
    "                    a = pathway[pathway[\"pathway_4\"] == j].index.tolist()\n",
    "                    if len(a)>0:\n",
    "                        temp_functions.append(a[0])\n",
    "    for k in temp_functions:\n",
    "        counter = counter + 1\n",
    "        temp_index = functions.index(k)\n",
    "        final_matrix[temp_index] = [final_matrix[temp_index][n] + temp_values[n] for n in range(len(temp_values))] \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d39c28a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "final_data_functions_by_drugs = pd.DataFrame(final_matrix,index = functions, columns=drugs)\n",
    "final_data_functions_by_drugs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5d43dda7",
   "metadata": {},
   "outputs": [],
   "source": [
    "filepath = Path('/Users/louxuwen/Desktop/Documents/GitHub/SP23-BENG213/Project/Map Gene to KEGG Pathway/Final_data_functions_by_drugs.csv')  \n",
    "filepath.parent.mkdir(parents=True, exist_ok=True)  \n",
    "final_data_functions_by_drugs.to_csv(filepath)  "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e4a4e2cc",
   "metadata": {},
   "source": [
    "## establish matrix as pathway by drugs ##"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5d74611a",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
