{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Homework 0: Optional Intro to Python Data Science\n",
    "\n",
    "This homework is an optional primer to how we will use python in this class. \n",
    "\n",
    "Sections 1-4 simply describe the fundamental Python concepts you should be familiar with, and link to external tutorials that may be valuable. There are empty code cells below each section for you to try out what you learn in the linked tutorials.\n",
    "\n",
    "When you reach section 5, the tutorial becomes more hands-on and specific to this course, with useful coding examples included."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "toc": true
   },
   "source": [
    "<h1>Table of Contents<span class=\"tocSkip\"></span></h1>\n",
    "<div class=\"toc\"><ul class=\"toc-item\"><li><span><a href=\"#Helpful-Resources\" data-toc-modified-id=\"Helpful-Resources-1\"><span class=\"toc-item-num\">1&nbsp;&nbsp;</span>Helpful Resources</a></span></li><li><span><a href=\"#Basics\" data-toc-modified-id=\"Basics-2\"><span class=\"toc-item-num\">2&nbsp;&nbsp;</span>Basics</a></span><ul class=\"toc-item\"><li><span><a href=\"#Jupyter-Notebooks\" data-toc-modified-id=\"Jupyter-Notebooks-2.1\"><span class=\"toc-item-num\">2.1&nbsp;&nbsp;</span>Jupyter Notebooks</a></span></li><li><span><a href=\"#Basic-Coding\" data-toc-modified-id=\"Basic-Coding-2.2\"><span class=\"toc-item-num\">2.2&nbsp;&nbsp;</span>Basic Coding</a></span></li></ul></li><li><span><a href=\"#Basic-Data-Structures\" data-toc-modified-id=\"Basic-Data-Structures-3\"><span class=\"toc-item-num\">3&nbsp;&nbsp;</span>Basic Data Structures</a></span><ul class=\"toc-item\"><li><span><a href=\"#Lists\" data-toc-modified-id=\"Lists-3.1\"><span class=\"toc-item-num\">3.1&nbsp;&nbsp;</span>Lists</a></span></li><li><span><a href=\"#Dictionaries\" data-toc-modified-id=\"Dictionaries-3.2\"><span class=\"toc-item-num\">3.2&nbsp;&nbsp;</span>Dictionaries</a></span></li><li><span><a href=\"#Other-Groupings:-Strings,-Sets,-and-Tuples\" data-toc-modified-id=\"Other-Groupings:-Strings,-Sets,-and-Tuples-3.3\"><span class=\"toc-item-num\">3.3&nbsp;&nbsp;</span>Other Groupings: Strings, Sets, and Tuples</a></span></li><li><span><a href=\"#Other-Useful-Functions-to-Know\" data-toc-modified-id=\"Other-Useful-Functions-to-Know-3.4\"><span class=\"toc-item-num\">3.4&nbsp;&nbsp;</span>Other Useful Functions to Know</a></span></li></ul></li><li><span><a href=\"#NumPy\" data-toc-modified-id=\"NumPy-4\"><span class=\"toc-item-num\">4&nbsp;&nbsp;</span>NumPy</a></span><ul class=\"toc-item\"><li><span><a href=\"#What-Is-a-Python-Package?\" data-toc-modified-id=\"What-Is-a-Python-Package?-4.1\"><span class=\"toc-item-num\">4.1&nbsp;&nbsp;</span>What Is a Python Package?</a></span></li><li><span><a href=\"#Learn-NumPy\" data-toc-modified-id=\"Learn-NumPy-4.2\"><span class=\"toc-item-num\">4.2&nbsp;&nbsp;</span>Learn NumPy</a></span></li></ul></li><li><span><a href=\"#Pandas\" data-toc-modified-id=\"Pandas-5\"><span class=\"toc-item-num\">5&nbsp;&nbsp;</span>Pandas</a></span><ul class=\"toc-item\"><li><span><a href=\"#Reading-in-Your-Data\" data-toc-modified-id=\"Reading-in-Your-Data-5.1\"><span class=\"toc-item-num\">5.1&nbsp;&nbsp;</span>Reading in Your Data</a></span></li><li><span><a href=\"#View-Your-Data\" data-toc-modified-id=\"View-Your-Data-5.2\"><span class=\"toc-item-num\">5.2&nbsp;&nbsp;</span>View Your Data</a></span></li><li><span><a href=\"#Basic-Data-Access\" data-toc-modified-id=\"Basic-Data-Access-5.3\"><span class=\"toc-item-num\">5.3&nbsp;&nbsp;</span>Basic Data Access</a></span><ul class=\"toc-item\"><li><span><a href=\"#Columns\" data-toc-modified-id=\"Columns-5.3.1\"><span class=\"toc-item-num\">5.3.1&nbsp;&nbsp;</span>Columns</a></span></li><li><span><a href=\"#Rows\" data-toc-modified-id=\"Rows-5.3.2\"><span class=\"toc-item-num\">5.3.2&nbsp;&nbsp;</span>Rows</a></span></li><li><span><a href=\"#Single-Values-and-Slicing-Both-Ways-at-Once\" data-toc-modified-id=\"Single-Values-and-Slicing-Both-Ways-at-Once-5.3.3\"><span class=\"toc-item-num\">5.3.3&nbsp;&nbsp;</span>Single Values and Slicing Both Ways at Once</a></span></li><li><span><a href=\"#Other-Ways-of-Indexing\" data-toc-modified-id=\"Other-Ways-of-Indexing-5.3.4\"><span class=\"toc-item-num\">5.3.4&nbsp;&nbsp;</span>Other Ways of Indexing</a></span></li></ul></li><li><span><a href=\"#Data-Access:-Using-the-Sample-Table\" data-toc-modified-id=\"Data-Access:-Using-the-Sample-Table-5.4\"><span class=\"toc-item-num\">5.4&nbsp;&nbsp;</span>Data Access: Using the Sample Table</a></span><ul class=\"toc-item\"><li><span><a href=\"#Unique()\" data-toc-modified-id=\"Unique()-5.4.1\"><span class=\"toc-item-num\">5.4.1&nbsp;&nbsp;</span>Unique()</a></span></li><li><span><a href=\"#Boolean-Series\" data-toc-modified-id=\"Boolean-Series-5.4.2\"><span class=\"toc-item-num\">5.4.2&nbsp;&nbsp;</span>Boolean Series</a></span></li></ul></li><li><span><a href=\"#Basic-Statistics,-Math,-and-Sorting:-Glu-v-Fru-Example\" data-toc-modified-id=\"Basic-Statistics,-Math,-and-Sorting:-Glu-v-Fru-Example-5.5\"><span class=\"toc-item-num\">5.5&nbsp;&nbsp;</span>Basic Statistics, Math, and Sorting: Glu v Fru Example</a></span><ul class=\"toc-item\"><li><span><a href=\"#Results-DataFrame-and-DataFrame-Means\" data-toc-modified-id=\"Results-DataFrame-and-DataFrame-Means-5.5.1\"><span class=\"toc-item-num\">5.5.1&nbsp;&nbsp;</span>Results DataFrame and DataFrame Means</a></span></li><li><span><a href=\"#Subtraction-and-Absolute-Value\" data-toc-modified-id=\"Subtraction-and-Absolute-Value-5.5.2\"><span class=\"toc-item-num\">5.5.2&nbsp;&nbsp;</span>Subtraction and Absolute Value</a></span></li><li><span><a href=\"#Sort-the-DataFrame\" data-toc-modified-id=\"Sort-the-DataFrame-5.5.3\"><span class=\"toc-item-num\">5.5.3&nbsp;&nbsp;</span>Sort the DataFrame</a></span></li><li><span><a href=\"#Connecting-to-Reality:-Adding-Gene-Information\" data-toc-modified-id=\"Connecting-to-Reality:-Adding-Gene-Information-5.5.4\"><span class=\"toc-item-num\">5.5.4&nbsp;&nbsp;</span>Connecting to Reality: Adding Gene Information</a></span></li></ul></li></ul></li></ul></div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Helpful Resources\n",
    "\n",
    "If you have never coded or used Python before, you should browse the following resources:\n",
    "\n",
    "- [The official python tutorial](https://docs.python.org/3/tutorial/)\n",
    "- [An introduction to programming for bioscientists](https://journals.plos.org/ploscompbiol/article/file?id=10.1371/journal.pcbi.1004867&type=printable) (easy to read and follow along)\n",
    "- [W3schools, an excellent interactive tutorial](https://www.w3schools.com/python/default.asp)\n",
    "\n",
    "The number one most valuable skill for programming is - surprisingly - knowing how to google exactly what you need. Most questions are already answered on Stack Overflow or in package documentation, but it there is a learning curve to finding and re-using this information correctly. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Basics\n",
    "\n",
    "### Jupyter Notebooks\n",
    "\n",
    "[Learn about Jupyter Notebooks.](https://www.dataquest.io/blog/jupyter-notebook-tutorial/)\n",
    "\n",
    "We will use two main types of cells: Markdown, which contains headings and paragraphs, and Code, which contains code. Switch types by selecting a cell and navigating to Cell -> Cell Type in the top toolbar. \n",
    "\n",
    "To run a code block, either click \"Run\" in the top toolbar or press Ctrl+Enter while typing in that block. \n",
    "\n",
    "All code cells have \"In \\[x\\]:\" to their left. While the cell is running, a star will appear in the brackets. If the code has already ran, the number in the bracket will count how many cells have been ran since the kernel started.\n",
    "\n",
    "If jupyter notebook is misbehaving (unusual errors, very slow response times), try restarting the kernel by navigating to Kernel -> Restart in the top toolbar. All variables will be lost when you do this."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Basic Coding\n",
    "\n",
    "Be familiar with [if statements](https://www.w3schools.com/python/python_conditions.asp), [while](https://www.w3schools.com/python/python_while_loops.asp) and [for](https://www.w3schools.com/python/python_for_loops.asp) loops, and [functions](https://www.w3schools.com/python/python_functions.asp)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Basic Data Structures\n",
    "\n",
    "This section introduces the basic data structures built into python. You should be broadly familiar with what they are,\n",
    "\n",
    "### Lists\n",
    "\n",
    "[Learn the basics of lists.](https://docs.python.org/3/tutorial/introduction.html#lists)\n",
    "\n",
    "[Understand the range function.](https://docs.python.org/3/tutorial/controlflow.html#the-range-function)\n",
    "\n",
    "[More advanced: Try out pythonic list comprehensions.](https://docs.python.org/3/tutorial/datastructures.html#list-comprehensions)\n",
    "\n",
    "Lists are the simplest way to store a group of numbers in order in python. In BENG 212, we will tend to use more advanced data structures for storing our main data, but it is important to understand them because other data structures are built off of them. Lists themselves also are useful for two main reasons:\n",
    "\n",
    "- If we need to store a simple set of numbers with no labels and no need for complex mathematics, the list syntax is the most convenient way to do so.\n",
    "- Interfacing with functions, such as those for statistics and supervised learning, may require lists and will often return their output as lists.\n",
    "\n",
    "Try out some of the examples from the links above in the code block below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Try making lists, growing them, and iterating through them! (use the links as a guide)\n",
    "my_list = []"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Dictionaries\n",
    "\n",
    "[Try out a dictionary tutorial.](https://www.w3schools.com/python/python_dictionaries_access.asp)\n",
    "\n",
    "Dictionaries are a collection of items, where each item has a unique key (like the words in a dictionary, what you \"look up\") and value (like the definition in a dictionary). These are useful for mapping one thing to another; for example, you could use a dictionary to assign a color to each gene category, then look up that color with your dictionary as you plot. For now, just be comfortable creating a dictionary, accessing items, and adding new items."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Other Groupings: Strings, Sets, and Tuples\n",
    "\n",
    "[Strings](https://www.w3schools.com/python/python_strings.asp) are essentially lists of characters used for names and words. There are many useful built-in string functions for operations like capitalization and searching for substrings.\n",
    "\n",
    "[Sets](https://docs.python.org/3/tutorial/datastructures.html#sets) are unordered lists that can't contain repeat elements. They are the most useful when we want to perform set operations like union and intersection. For example, you might have two sets of important genes, and want to know [which genes appear in both sets](https://www.w3schools.com/python/ref_set_intersection.asp). You can convert between sets and lists using their respective constructors (set() and list()).\n",
    "\n",
    "[Tuples](https://docs.python.org/3/library/stdtypes.html#tuples) are the least commonly used sequence types. They use parentheses instead of brackets and cannot be modified the way that lists can. Don't worry about them too much, but be aware that they can be converted back to lists using the list() constructor.\n",
    "\n",
    "For each of these, know how they are constructed and what they can be used for. When you run into an application for them in a problem or you find them as an output to a function, you can always pull up the documentation for them."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Other Useful Functions to Know\n",
    "\n",
    "- [type()](https://www.programiz.com/python-programming/methods/built-in/type): Returns the type of a variable, which is useful for debugging and understanding code.\n",
    "- [isinstance()](https://www.programiz.com/python-programming/methods/built-in/isinstance): If you want to do something different to a variable depending on its type, you could create an if statement using this function. For example, you might have a list that combines numbers and strings, and you want to round all of the numbers to a specific number of decimals but do nothing to the strings.\n",
    "- [zip()](https://www.geeksforgeeks.org/zip-in-python/#highlighter_116733): If you are iterating through two lists at once, you can zip them together, which creates a tuple-like object.\n",
    "- [enumerate()](https://www.geeksforgeeks.org/enumerate-in-python/): Similarly, if you are iterating through a list and also want the number of iterations, you can use this function. This is the same as zip(range(len(my_list)), my_list)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## NumPy\n",
    "\n",
    "### What Is a Python Package?\n",
    "\n",
    "This is the first package we will introduce. Packages are made by developers to provide new classes and functions and empower you as a data scientist. You must have packages [installed on your computer](https://docs.python.org/3/installing/index.html) to use them, which can be done using the command line and a program called \"pip\". Numpy should have been included in your python installation, but other packages might not. The easiest way to check if a package is on your computer is to try to import it and see if it can't be found.\n",
    "\n",
    "To bring a package into your notebook workspace, you use [the import command](https://realpython.com/python-import/). You can use the word \"from\" to specify parts of a package if you only need a function or two, and you can use the word \"as\" to give a package a nickname. It is pretty common to import numpy as np.\n",
    "\n",
    "### Learn NumPy\n",
    "\n",
    "NumPy is known for its arrays, which are like lists but support arithmetic and a ton of other functionality. It also has many other math functions, like random number generation, trigonometry, and tons more. This places it at the core of a lot of data science work in Python.\n",
    "\n",
    "However, the drawback to NumPy is that it doesn't support labeling of rows and columns for its arrays the way that another package, Pandas, does. I encourage you to know how to use NumPy, but switch to Pandas for data storage and manipulation whenever possible in order to maintain smarter indexing. Under the hood, Pandas uses NumPy arrays - it is therefore important to know both and the distinction between them.\n",
    "\n",
    "Follow along with this [excellent illustrated guide to NumPy](https://medium.com/better-programming/numpy-illustrated-the-visual-guide-to-numpy-3b1d4976de1d), or try the [W3schools tutorial](https://www.w3schools.com/python/numpy_intro.asp)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "\n",
    "# try creating arrays and playing around with them"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Pandas\n",
    "\n",
    "This is the most useful package for this class, as it is the industry standard for storing labeled data. Check out [this tutorial](https://pandas.pydata.org/pandas-docs/stable/user_guide/10min.html) and the rest of the Pandas User Guide.\n",
    "\n",
    "There are three main classes to be aware of:\n",
    "- [pd.DataFrame](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.html?highlight=dataframe#pandas.DataFrame): This is a (usually) two-dimensional array with labels for the rows (index) and columns (columns). We'll use this most often.\n",
    "- [pd.Series](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.html?highlight=series#pandas.Series): This is the one-dimensional version of a DataFrame, which only has rows (index) and a single column name.\n",
    "- [pd.Index](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Index.html?highlight=index#pandas.Index): The axis labels (rows and columns of DataFrames, rows of Series) are technically stored as Index datatypes. You won't need to work with this very often since manipulating the other two classes will automatically update their associated indices.\n",
    "\n",
    "### Reading in Your Data\n",
    "\n",
    "The majority of the homeworks in this class will use the same E. coli RNA sequencing dataset, which we call PRECISE 1 and was the subject of [this paper](https://www.nature.com/articles/s41467-019-13483-w). Download it from Canvas and save it in the same folder as this notebook (or modify the path below to point to your preferred location).\n",
    "\n",
    "Open the file on your computer just to see what's inside. This is an excel file with several sheets. The first sheet, 'Summary', is a table of contents for the rest of the file. We are interested in the 'Expression Data'; we will learn the tools for analyzing this in class. We are also interested in the 'Experimental Conditions', since the only way to interpret our results is to understand what was done to the cells to induce the gene expression we observe. Finally, the 'Gene Information' sheet describes each gene, so it will help us interpret our results, too. Let's read in all three.\n",
    "\n",
    "When you see a new function the first time, I recommend finding its documentation to learn exactly what it is doing and how else it can be used. [This is the Pandas function for reading excel files used below.](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_excel.html) The toolbar on the left of the pandas documentation shows related functions, so you can also learn how to read in other file types. Note that we had to specify sheet names because the file contains multiple sheets, and we use 'index_col = 0' because the labels are in the 0th column of the sheets in all cases.\n",
    "\n",
    "Pro tip: You can also view the documentation by pressing 'Shift + Tab' inside the parentheses of the function. This will only work if the package it's from has previously been imported. Use this any time you forget the order of inputs. [There are tons of useful jupyter shortcuts like this one.](https://medium.com/@rrfd/jupyter-notebook-shortcuts-tips-and-tricks-top-nbextensions-bring-order-to-your-notebooks-pt-1-b58f270963f9)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import the pandas package\n",
    "import pandas as pd\n",
    "\n",
    "# read in the data\n",
    "data = pd.read_excel('Homework_Data.xlsx', \n",
    "                     sheet_name = 'Expression Data',\n",
    "                     index_col = 0)\n",
    "\n",
    "sample_table = pd.read_excel('Homework_Data.xlsx',\n",
    "                             sheet_name = 'Experimental Conditions',\n",
    "                             index_col = 0)\n",
    "\n",
    "gene_table = pd.read_excel('Homework_Data.xlsx',\n",
    "                           sheet_name = 'Gene Information',\n",
    "                           index_col = 0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### View Your Data\n",
    "\n",
    "To view an output in python, you have three options:\n",
    "\n",
    "- return the output in the last code line of the cell. Under the hood, this calls display().\n",
    "- print() it out: this is fast and produces a simple output, preferred for most data types\n",
    "- display() it: this is preferred for pandas because it formats everything better\n",
    "\n",
    "So, if you just want to see a DataFrame, you can use the first option and type its name in the cell (as I've done below). If you are running more complex code and want to display it in the middle of a cell execution, then you should use option 3.\n",
    "\n",
    "Note that the shape of the DataFrame is viewable in the bottom left of these outputs - you could also show only the shape using ['data.shape'](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.shape.html). By default, pandas only shows 10 rows and 20 columns in these outputs. You can modify those settings to show more of the DataFrame using a command like [pd.set_option('display.max_rows', 200)](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.set_option.html?highlight=set_option#pandas.set_option). You can crash your browser by outputting too much data, however."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>base__wt_glc__1</th>\n",
       "      <th>base__wt_glc__2</th>\n",
       "      <th>fur__wt_dpd__1</th>\n",
       "      <th>fur__wt_dpd__2</th>\n",
       "      <th>fur__wt_fe__1</th>\n",
       "      <th>fur__wt_fe__2</th>\n",
       "      <th>fur__delfur_dpd__1</th>\n",
       "      <th>fur__delfur_dpd__2</th>\n",
       "      <th>fur__delfur_fe2__1</th>\n",
       "      <th>fur__delfur_fe2__2</th>\n",
       "      <th>...</th>\n",
       "      <th>rpoB__wt_lb__2</th>\n",
       "      <th>rpoB__rpoBE546V_031__1</th>\n",
       "      <th>rpoB__rpoBE546V_044__1</th>\n",
       "      <th>rpoB__rpoBE546V_lb__1</th>\n",
       "      <th>rpoB__rpoBE546V_lb__2</th>\n",
       "      <th>rpoB__rpoBE546V_glc__1</th>\n",
       "      <th>rpoB__rpoBE672K_glc__1</th>\n",
       "      <th>rpoB__rpoBE672K_glc__2</th>\n",
       "      <th>rpoB__rpoBE672K_lb__1</th>\n",
       "      <th>rpoB__rpoBE672K_lb__2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>b0002</th>\n",
       "      <td>11.006777</td>\n",
       "      <td>11.130303</td>\n",
       "      <td>11.705085</td>\n",
       "      <td>11.888364</td>\n",
       "      <td>11.064951</td>\n",
       "      <td>10.779232</td>\n",
       "      <td>9.976561</td>\n",
       "      <td>10.291286</td>\n",
       "      <td>11.229957</td>\n",
       "      <td>11.214245</td>\n",
       "      <td>...</td>\n",
       "      <td>7.489847</td>\n",
       "      <td>10.188248</td>\n",
       "      <td>10.304740</td>\n",
       "      <td>8.177974</td>\n",
       "      <td>8.150298</td>\n",
       "      <td>10.927531</td>\n",
       "      <td>11.000981</td>\n",
       "      <td>10.999663</td>\n",
       "      <td>8.136954</td>\n",
       "      <td>8.154207</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0003</th>\n",
       "      <td>10.415779</td>\n",
       "      <td>10.523245</td>\n",
       "      <td>11.423968</td>\n",
       "      <td>11.803928</td>\n",
       "      <td>10.777126</td>\n",
       "      <td>10.597971</td>\n",
       "      <td>9.596992</td>\n",
       "      <td>10.191653</td>\n",
       "      <td>10.898128</td>\n",
       "      <td>10.861337</td>\n",
       "      <td>...</td>\n",
       "      <td>8.002871</td>\n",
       "      <td>9.737415</td>\n",
       "      <td>10.019916</td>\n",
       "      <td>8.108515</td>\n",
       "      <td>8.146578</td>\n",
       "      <td>10.293188</td>\n",
       "      <td>10.639809</td>\n",
       "      <td>10.638433</td>\n",
       "      <td>7.746108</td>\n",
       "      <td>7.673084</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0004</th>\n",
       "      <td>10.594996</td>\n",
       "      <td>10.725169</td>\n",
       "      <td>10.457404</td>\n",
       "      <td>10.779308</td>\n",
       "      <td>10.395113</td>\n",
       "      <td>10.114110</td>\n",
       "      <td>8.741777</td>\n",
       "      <td>9.082381</td>\n",
       "      <td>10.185341</td>\n",
       "      <td>10.164834</td>\n",
       "      <td>...</td>\n",
       "      <td>8.006288</td>\n",
       "      <td>9.916193</td>\n",
       "      <td>10.047979</td>\n",
       "      <td>8.034197</td>\n",
       "      <td>8.096981</td>\n",
       "      <td>10.247070</td>\n",
       "      <td>10.876158</td>\n",
       "      <td>10.882020</td>\n",
       "      <td>7.793896</td>\n",
       "      <td>7.822367</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0005</th>\n",
       "      <td>6.316225</td>\n",
       "      <td>6.258604</td>\n",
       "      <td>5.422259</td>\n",
       "      <td>5.336264</td>\n",
       "      <td>6.716210</td>\n",
       "      <td>6.411023</td>\n",
       "      <td>4.627102</td>\n",
       "      <td>4.756298</td>\n",
       "      <td>6.527841</td>\n",
       "      <td>6.136345</td>\n",
       "      <td>...</td>\n",
       "      <td>2.005866</td>\n",
       "      <td>6.028688</td>\n",
       "      <td>6.221017</td>\n",
       "      <td>3.292041</td>\n",
       "      <td>4.404625</td>\n",
       "      <td>7.551864</td>\n",
       "      <td>7.014460</td>\n",
       "      <td>7.019091</td>\n",
       "      <td>3.622154</td>\n",
       "      <td>4.223552</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0006</th>\n",
       "      <td>6.915894</td>\n",
       "      <td>6.897703</td>\n",
       "      <td>6.775777</td>\n",
       "      <td>6.782750</td>\n",
       "      <td>6.761954</td>\n",
       "      <td>6.816691</td>\n",
       "      <td>6.686925</td>\n",
       "      <td>6.860185</td>\n",
       "      <td>6.862335</td>\n",
       "      <td>6.817658</td>\n",
       "      <td>...</td>\n",
       "      <td>6.606902</td>\n",
       "      <td>6.958569</td>\n",
       "      <td>7.131878</td>\n",
       "      <td>7.011375</td>\n",
       "      <td>6.916578</td>\n",
       "      <td>6.882943</td>\n",
       "      <td>6.958771</td>\n",
       "      <td>7.071100</td>\n",
       "      <td>6.903703</td>\n",
       "      <td>6.795143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b4688</th>\n",
       "      <td>2.473267</td>\n",
       "      <td>2.498651</td>\n",
       "      <td>0.798125</td>\n",
       "      <td>0.823918</td>\n",
       "      <td>2.424054</td>\n",
       "      <td>2.138454</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.741850</td>\n",
       "      <td>2.637325</td>\n",
       "      <td>2.687802</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.564523</td>\n",
       "      <td>0.848379</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.175645</td>\n",
       "      <td>2.357293</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b4693</th>\n",
       "      <td>0.868970</td>\n",
       "      <td>1.426800</td>\n",
       "      <td>2.509295</td>\n",
       "      <td>2.168248</td>\n",
       "      <td>1.756031</td>\n",
       "      <td>2.136488</td>\n",
       "      <td>3.706387</td>\n",
       "      <td>3.290683</td>\n",
       "      <td>4.268871</td>\n",
       "      <td>4.252881</td>\n",
       "      <td>...</td>\n",
       "      <td>1.129557</td>\n",
       "      <td>2.490407</td>\n",
       "      <td>2.211672</td>\n",
       "      <td>2.490141</td>\n",
       "      <td>2.739155</td>\n",
       "      <td>2.260636</td>\n",
       "      <td>1.320209</td>\n",
       "      <td>1.796015</td>\n",
       "      <td>1.989279</td>\n",
       "      <td>1.731708</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b4696_1</th>\n",
       "      <td>2.646606</td>\n",
       "      <td>2.545537</td>\n",
       "      <td>3.762534</td>\n",
       "      <td>3.639483</td>\n",
       "      <td>2.064642</td>\n",
       "      <td>2.014469</td>\n",
       "      <td>3.510175</td>\n",
       "      <td>3.327276</td>\n",
       "      <td>2.468861</td>\n",
       "      <td>2.432075</td>\n",
       "      <td>...</td>\n",
       "      <td>2.407559</td>\n",
       "      <td>2.731006</td>\n",
       "      <td>2.443753</td>\n",
       "      <td>1.802541</td>\n",
       "      <td>2.257490</td>\n",
       "      <td>2.402256</td>\n",
       "      <td>1.948501</td>\n",
       "      <td>1.554729</td>\n",
       "      <td>2.326432</td>\n",
       "      <td>1.807898</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b4696_2</th>\n",
       "      <td>3.084833</td>\n",
       "      <td>3.148125</td>\n",
       "      <td>3.902075</td>\n",
       "      <td>3.997867</td>\n",
       "      <td>2.639225</td>\n",
       "      <td>2.200407</td>\n",
       "      <td>3.954130</td>\n",
       "      <td>3.917911</td>\n",
       "      <td>3.044833</td>\n",
       "      <td>3.115995</td>\n",
       "      <td>...</td>\n",
       "      <td>2.638285</td>\n",
       "      <td>3.124999</td>\n",
       "      <td>3.258820</td>\n",
       "      <td>3.227410</td>\n",
       "      <td>2.718594</td>\n",
       "      <td>2.406698</td>\n",
       "      <td>3.096639</td>\n",
       "      <td>3.117403</td>\n",
       "      <td>1.913497</td>\n",
       "      <td>2.105586</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b4705</th>\n",
       "      <td>6.192737</td>\n",
       "      <td>4.744070</td>\n",
       "      <td>1.118211</td>\n",
       "      <td>1.150873</td>\n",
       "      <td>4.720938</td>\n",
       "      <td>4.211397</td>\n",
       "      <td>2.160077</td>\n",
       "      <td>1.046398</td>\n",
       "      <td>2.788759</td>\n",
       "      <td>3.595742</td>\n",
       "      <td>...</td>\n",
       "      <td>2.156118</td>\n",
       "      <td>1.351163</td>\n",
       "      <td>2.264639</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>5.469045</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3887 rows Ã— 102 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         base__wt_glc__1  base__wt_glc__2  fur__wt_dpd__1  fur__wt_dpd__2  \\\n",
       "b0002          11.006777        11.130303       11.705085       11.888364   \n",
       "b0003          10.415779        10.523245       11.423968       11.803928   \n",
       "b0004          10.594996        10.725169       10.457404       10.779308   \n",
       "b0005           6.316225         6.258604        5.422259        5.336264   \n",
       "b0006           6.915894         6.897703        6.775777        6.782750   \n",
       "...                  ...              ...             ...             ...   \n",
       "b4688           2.473267         2.498651        0.798125        0.823918   \n",
       "b4693           0.868970         1.426800        2.509295        2.168248   \n",
       "b4696_1         2.646606         2.545537        3.762534        3.639483   \n",
       "b4696_2         3.084833         3.148125        3.902075        3.997867   \n",
       "b4705           6.192737         4.744070        1.118211        1.150873   \n",
       "\n",
       "         fur__wt_fe__1  fur__wt_fe__2  fur__delfur_dpd__1  fur__delfur_dpd__2  \\\n",
       "b0002        11.064951      10.779232            9.976561           10.291286   \n",
       "b0003        10.777126      10.597971            9.596992           10.191653   \n",
       "b0004        10.395113      10.114110            8.741777            9.082381   \n",
       "b0005         6.716210       6.411023            4.627102            4.756298   \n",
       "b0006         6.761954       6.816691            6.686925            6.860185   \n",
       "...                ...            ...                 ...                 ...   \n",
       "b4688         2.424054       2.138454            0.000000            0.741850   \n",
       "b4693         1.756031       2.136488            3.706387            3.290683   \n",
       "b4696_1       2.064642       2.014469            3.510175            3.327276   \n",
       "b4696_2       2.639225       2.200407            3.954130            3.917911   \n",
       "b4705         4.720938       4.211397            2.160077            1.046398   \n",
       "\n",
       "         fur__delfur_fe2__1  fur__delfur_fe2__2  ...  rpoB__wt_lb__2  \\\n",
       "b0002             11.229957           11.214245  ...        7.489847   \n",
       "b0003             10.898128           10.861337  ...        8.002871   \n",
       "b0004             10.185341           10.164834  ...        8.006288   \n",
       "b0005              6.527841            6.136345  ...        2.005866   \n",
       "b0006              6.862335            6.817658  ...        6.606902   \n",
       "...                     ...                 ...  ...             ...   \n",
       "b4688              2.637325            2.687802  ...        0.000000   \n",
       "b4693              4.268871            4.252881  ...        1.129557   \n",
       "b4696_1            2.468861            2.432075  ...        2.407559   \n",
       "b4696_2            3.044833            3.115995  ...        2.638285   \n",
       "b4705              2.788759            3.595742  ...        2.156118   \n",
       "\n",
       "         rpoB__rpoBE546V_031__1  rpoB__rpoBE546V_044__1  \\\n",
       "b0002                 10.188248               10.304740   \n",
       "b0003                  9.737415               10.019916   \n",
       "b0004                  9.916193               10.047979   \n",
       "b0005                  6.028688                6.221017   \n",
       "b0006                  6.958569                7.131878   \n",
       "...                         ...                     ...   \n",
       "b4688                  1.564523                0.848379   \n",
       "b4693                  2.490407                2.211672   \n",
       "b4696_1                2.731006                2.443753   \n",
       "b4696_2                3.124999                3.258820   \n",
       "b4705                  1.351163                2.264639   \n",
       "\n",
       "         rpoB__rpoBE546V_lb__1  rpoB__rpoBE546V_lb__2  rpoB__rpoBE546V_glc__1  \\\n",
       "b0002                 8.177974               8.150298               10.927531   \n",
       "b0003                 8.108515               8.146578               10.293188   \n",
       "b0004                 8.034197               8.096981               10.247070   \n",
       "b0005                 3.292041               4.404625                7.551864   \n",
       "b0006                 7.011375               6.916578                6.882943   \n",
       "...                        ...                    ...                     ...   \n",
       "b4688                 0.000000               0.000000                3.175645   \n",
       "b4693                 2.490141               2.739155                2.260636   \n",
       "b4696_1               1.802541               2.257490                2.402256   \n",
       "b4696_2               3.227410               2.718594                2.406698   \n",
       "b4705                 0.000000               0.000000                5.469045   \n",
       "\n",
       "         rpoB__rpoBE672K_glc__1  rpoB__rpoBE672K_glc__2  \\\n",
       "b0002                 11.000981               10.999663   \n",
       "b0003                 10.639809               10.638433   \n",
       "b0004                 10.876158               10.882020   \n",
       "b0005                  7.014460                7.019091   \n",
       "b0006                  6.958771                7.071100   \n",
       "...                         ...                     ...   \n",
       "b4688                  2.357293                0.000000   \n",
       "b4693                  1.320209                1.796015   \n",
       "b4696_1                1.948501                1.554729   \n",
       "b4696_2                3.096639                3.117403   \n",
       "b4705                  0.000000                0.000000   \n",
       "\n",
       "         rpoB__rpoBE672K_lb__1  rpoB__rpoBE672K_lb__2  \n",
       "b0002                 8.136954               8.154207  \n",
       "b0003                 7.746108               7.673084  \n",
       "b0004                 7.793896               7.822367  \n",
       "b0005                 3.622154               4.223552  \n",
       "b0006                 6.903703               6.795143  \n",
       "...                        ...                    ...  \n",
       "b4688                 0.000000               0.000000  \n",
       "b4693                 1.989279               1.731708  \n",
       "b4696_1               2.326432               1.807898  \n",
       "b4696_2               1.913497               2.105586  \n",
       "b4705                 0.000000               0.000000  \n",
       "\n",
       "[3887 rows x 102 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Shown above is the main dataset. We will learn how to analyze and visualize it throughout this class and the homeworks. You'll notice that:\n",
    "\n",
    "- The rows correspond to genes. They are labeled by locus tag, which in E. coli is also called the \"b number\". The RNA sequencing pipeline counts nearly all the coding genes and pseudogenes, resulting in 3887 rows.\n",
    "- The columns correspond to samples. Each column is a population of cells grown under a specific condition that has been RNA sequenced. These have a name that looks like xxx__yyyyy__z, where:\n",
    "    - xxx (the part before the first double underscore) is the study or project name (largest grouping)\n",
    "    - yyyyy (middle portion) is the condition name\n",
    "    - z is the replicate number\n",
    "- The elements are [log(TPM) values](https://rna-seqblog.com/rpkm-fpkm-and-tpm-clearly-explained/). They are nonnegative decimal numbers corresponding to how much the gene was expressed in the given condition. TPM refers to Transcripts Per Million total transcripts, so each column sums to approximately the same number. Gene expression tends to vary on a logarithmic scale, so the log has been taken in order to compare more linearly.\n",
    "\n",
    "**Important Note:** In this class, we have genes as rows and columns as conditions, but some data scientists prefer to use the [transpose](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.transpose.html) of this. When we move onto using some machine learning built-in functions, you will need to be prepared to transpose the dataset for using those functions. Having informative labels for the genes and conditions will help you to keep track of this."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Study</th>\n",
       "      <th>Condition ID</th>\n",
       "      <th>Replicate #</th>\n",
       "      <th>Strain Description</th>\n",
       "      <th>Base Media</th>\n",
       "      <th>Carbon Source (g/L)</th>\n",
       "      <th>Nitrogen Source (g/L)</th>\n",
       "      <th>Electron Acceptor</th>\n",
       "      <th>Trace Element Mixture</th>\n",
       "      <th>Supplement</th>\n",
       "      <th>Temperature (C)</th>\n",
       "      <th>pH</th>\n",
       "      <th>Culture Type</th>\n",
       "      <th>Growth Rate (1/hr)</th>\n",
       "      <th>Sequencing Machine</th>\n",
       "      <th>Additional Details</th>\n",
       "      <th>Biological Replicates</th>\n",
       "      <th>DOI</th>\n",
       "      <th>GEO</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>base__wt_glc__1</th>\n",
       "      <td>Base</td>\n",
       "      <td>wt_glc</td>\n",
       "      <td>1</td>\n",
       "      <td>Escherichia coli K-12 MG1655</td>\n",
       "      <td>M9</td>\n",
       "      <td>glucose(2)</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>O2</td>\n",
       "      <td>sauer trace element mixture</td>\n",
       "      <td>NaN</td>\n",
       "      <td>37</td>\n",
       "      <td>7.0</td>\n",
       "      <td>Batch</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MiSeq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>doi.org/10.1101/080929</td>\n",
       "      <td>GSE65643</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>base__wt_glc__2</th>\n",
       "      <td>Base</td>\n",
       "      <td>wt_glc</td>\n",
       "      <td>2</td>\n",
       "      <td>Escherichia coli K-12 MG1655</td>\n",
       "      <td>M9</td>\n",
       "      <td>glucose(2)</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>O2</td>\n",
       "      <td>sauer trace element mixture</td>\n",
       "      <td>NaN</td>\n",
       "      <td>37</td>\n",
       "      <td>7.0</td>\n",
       "      <td>Batch</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MiSeq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>doi.org/10.1101/080929</td>\n",
       "      <td>GSE65643</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>fur__wt_dpd__1</th>\n",
       "      <td>Fur</td>\n",
       "      <td>wt_dpd</td>\n",
       "      <td>1</td>\n",
       "      <td>Escherichia coli K-12 MG1655</td>\n",
       "      <td>M9</td>\n",
       "      <td>glucose(2)</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>O2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DPD(0.2mM)</td>\n",
       "      <td>37</td>\n",
       "      <td>7.0</td>\n",
       "      <td>Batch</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>MiSeq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>doi.org/10.1038/ncomms5910</td>\n",
       "      <td>GSE54900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>fur__wt_dpd__2</th>\n",
       "      <td>Fur</td>\n",
       "      <td>wt_dpd</td>\n",
       "      <td>2</td>\n",
       "      <td>Escherichia coli K-12 MG1655</td>\n",
       "      <td>M9</td>\n",
       "      <td>glucose(2)</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>O2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DPD(0.2mM)</td>\n",
       "      <td>37</td>\n",
       "      <td>7.0</td>\n",
       "      <td>Batch</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>MiSeq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>doi.org/10.1038/ncomms5910</td>\n",
       "      <td>GSE54900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>fur__wt_fe__1</th>\n",
       "      <td>Fur</td>\n",
       "      <td>wt_fe</td>\n",
       "      <td>1</td>\n",
       "      <td>Escherichia coli K-12 MG1655</td>\n",
       "      <td>M9</td>\n",
       "      <td>glucose(2)</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>O2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FeCl2(0.1mM)</td>\n",
       "      <td>37</td>\n",
       "      <td>7.0</td>\n",
       "      <td>Batch</td>\n",
       "      <td>1.060606</td>\n",
       "      <td>MiSeq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>doi.org/10.1038/ncomms5910</td>\n",
       "      <td>GSE54900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rpoB__rpoBE546V_glc__1</th>\n",
       "      <td>RpoB Knock-in</td>\n",
       "      <td>rpoBE546V_glc</td>\n",
       "      <td>1</td>\n",
       "      <td>Escherichia coli K-12 MG1655 rpoBE546V</td>\n",
       "      <td>M9</td>\n",
       "      <td>glucose(4)</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>O2</td>\n",
       "      <td>sauer trace element mixture</td>\n",
       "      <td>NaN</td>\n",
       "      <td>37</td>\n",
       "      <td>7.0</td>\n",
       "      <td>Batch</td>\n",
       "      <td>0.850000</td>\n",
       "      <td>MiSeq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>doi.org/10.1016/j.cels.2016.04.003</td>\n",
       "      <td>GSE59377</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rpoB__rpoBE672K_glc__1</th>\n",
       "      <td>RpoB Knock-in</td>\n",
       "      <td>rpoBE672K_glc</td>\n",
       "      <td>1</td>\n",
       "      <td>Escherichia coli K-12 MG1655 rpoBE672K</td>\n",
       "      <td>M9</td>\n",
       "      <td>glucose(4)</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>O2</td>\n",
       "      <td>sauer trace element mixture</td>\n",
       "      <td>NaN</td>\n",
       "      <td>37</td>\n",
       "      <td>7.0</td>\n",
       "      <td>Batch</td>\n",
       "      <td>0.880000</td>\n",
       "      <td>MiSeq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>doi.org/10.1016/j.cels.2016.04.003</td>\n",
       "      <td>GSE59377</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rpoB__rpoBE672K_glc__2</th>\n",
       "      <td>RpoB Knock-in</td>\n",
       "      <td>rpoBE672K_glc</td>\n",
       "      <td>2</td>\n",
       "      <td>Escherichia coli K-12 MG1655 rpoBE672K</td>\n",
       "      <td>M9</td>\n",
       "      <td>glucose(4)</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>O2</td>\n",
       "      <td>sauer trace element mixture</td>\n",
       "      <td>NaN</td>\n",
       "      <td>37</td>\n",
       "      <td>7.0</td>\n",
       "      <td>Batch</td>\n",
       "      <td>0.880000</td>\n",
       "      <td>MiSeq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>doi.org/10.1016/j.cels.2016.04.003</td>\n",
       "      <td>GSE59377</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rpoB__rpoBE672K_lb__1</th>\n",
       "      <td>RpoB Knock-in</td>\n",
       "      <td>rpoBE672K_lb</td>\n",
       "      <td>1</td>\n",
       "      <td>Escherichia coli K-12 MG1655 rpoBE672K</td>\n",
       "      <td>LB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>O2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>37</td>\n",
       "      <td>7.0</td>\n",
       "      <td>Batch</td>\n",
       "      <td>1.400000</td>\n",
       "      <td>MiSeq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>doi.org/10.1016/j.cels.2016.04.003</td>\n",
       "      <td>GSE59377</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rpoB__rpoBE672K_lb__2</th>\n",
       "      <td>RpoB Knock-in</td>\n",
       "      <td>rpoBE672K_lb</td>\n",
       "      <td>2</td>\n",
       "      <td>Escherichia coli K-12 MG1655 rpoBE672K</td>\n",
       "      <td>LB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>O2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>37</td>\n",
       "      <td>7.0</td>\n",
       "      <td>Batch</td>\n",
       "      <td>1.400000</td>\n",
       "      <td>MiSeq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>doi.org/10.1016/j.cels.2016.04.003</td>\n",
       "      <td>GSE59377</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>102 rows Ã— 19 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                Study   Condition ID  Replicate #  \\\n",
       "Sample ID                                                           \n",
       "base__wt_glc__1                  Base         wt_glc            1   \n",
       "base__wt_glc__2                  Base         wt_glc            2   \n",
       "fur__wt_dpd__1                    Fur         wt_dpd            1   \n",
       "fur__wt_dpd__2                    Fur         wt_dpd            2   \n",
       "fur__wt_fe__1                     Fur          wt_fe            1   \n",
       "...                               ...            ...          ...   \n",
       "rpoB__rpoBE546V_glc__1  RpoB Knock-in  rpoBE546V_glc            1   \n",
       "rpoB__rpoBE672K_glc__1  RpoB Knock-in  rpoBE672K_glc            1   \n",
       "rpoB__rpoBE672K_glc__2  RpoB Knock-in  rpoBE672K_glc            2   \n",
       "rpoB__rpoBE672K_lb__1   RpoB Knock-in   rpoBE672K_lb            1   \n",
       "rpoB__rpoBE672K_lb__2   RpoB Knock-in   rpoBE672K_lb            2   \n",
       "\n",
       "                                            Strain Description Base Media  \\\n",
       "Sample ID                                                                   \n",
       "base__wt_glc__1                   Escherichia coli K-12 MG1655         M9   \n",
       "base__wt_glc__2                   Escherichia coli K-12 MG1655         M9   \n",
       "fur__wt_dpd__1                    Escherichia coli K-12 MG1655         M9   \n",
       "fur__wt_dpd__2                    Escherichia coli K-12 MG1655         M9   \n",
       "fur__wt_fe__1                     Escherichia coli K-12 MG1655         M9   \n",
       "...                                                        ...        ...   \n",
       "rpoB__rpoBE546V_glc__1  Escherichia coli K-12 MG1655 rpoBE546V         M9   \n",
       "rpoB__rpoBE672K_glc__1  Escherichia coli K-12 MG1655 rpoBE672K         M9   \n",
       "rpoB__rpoBE672K_glc__2  Escherichia coli K-12 MG1655 rpoBE672K         M9   \n",
       "rpoB__rpoBE672K_lb__1   Escherichia coli K-12 MG1655 rpoBE672K         LB   \n",
       "rpoB__rpoBE672K_lb__2   Escherichia coli K-12 MG1655 rpoBE672K         LB   \n",
       "\n",
       "                       Carbon Source (g/L) Nitrogen Source (g/L)  \\\n",
       "Sample ID                                                          \n",
       "base__wt_glc__1                 glucose(2)              NH4Cl(1)   \n",
       "base__wt_glc__2                 glucose(2)              NH4Cl(1)   \n",
       "fur__wt_dpd__1                  glucose(2)              NH4Cl(1)   \n",
       "fur__wt_dpd__2                  glucose(2)              NH4Cl(1)   \n",
       "fur__wt_fe__1                   glucose(2)              NH4Cl(1)   \n",
       "...                                    ...                   ...   \n",
       "rpoB__rpoBE546V_glc__1          glucose(4)              NH4Cl(1)   \n",
       "rpoB__rpoBE672K_glc__1          glucose(4)              NH4Cl(1)   \n",
       "rpoB__rpoBE672K_glc__2          glucose(4)              NH4Cl(1)   \n",
       "rpoB__rpoBE672K_lb__1                  NaN                   NaN   \n",
       "rpoB__rpoBE672K_lb__2                  NaN                   NaN   \n",
       "\n",
       "                       Electron Acceptor        Trace Element Mixture  \\\n",
       "Sample ID                                                               \n",
       "base__wt_glc__1                       O2  sauer trace element mixture   \n",
       "base__wt_glc__2                       O2  sauer trace element mixture   \n",
       "fur__wt_dpd__1                        O2                          NaN   \n",
       "fur__wt_dpd__2                        O2                          NaN   \n",
       "fur__wt_fe__1                         O2                          NaN   \n",
       "...                                  ...                          ...   \n",
       "rpoB__rpoBE546V_glc__1                O2  sauer trace element mixture   \n",
       "rpoB__rpoBE672K_glc__1                O2  sauer trace element mixture   \n",
       "rpoB__rpoBE672K_glc__2                O2  sauer trace element mixture   \n",
       "rpoB__rpoBE672K_lb__1                 O2                          NaN   \n",
       "rpoB__rpoBE672K_lb__2                 O2                          NaN   \n",
       "\n",
       "                          Supplement  Temperature (C)   pH Culture Type  \\\n",
       "Sample ID                                                                 \n",
       "base__wt_glc__1                  NaN               37  7.0        Batch   \n",
       "base__wt_glc__2                  NaN               37  7.0        Batch   \n",
       "fur__wt_dpd__1            DPD(0.2mM)               37  7.0        Batch   \n",
       "fur__wt_dpd__2            DPD(0.2mM)               37  7.0        Batch   \n",
       "fur__wt_fe__1           FeCl2(0.1mM)               37  7.0        Batch   \n",
       "...                              ...              ...  ...          ...   \n",
       "rpoB__rpoBE546V_glc__1           NaN               37  7.0        Batch   \n",
       "rpoB__rpoBE672K_glc__1           NaN               37  7.0        Batch   \n",
       "rpoB__rpoBE672K_glc__2           NaN               37  7.0        Batch   \n",
       "rpoB__rpoBE672K_lb__1            NaN               37  7.0        Batch   \n",
       "rpoB__rpoBE672K_lb__2            NaN               37  7.0        Batch   \n",
       "\n",
       "                        Growth Rate (1/hr) Sequencing Machine  \\\n",
       "Sample ID                                                       \n",
       "base__wt_glc__1                        NaN              MiSeq   \n",
       "base__wt_glc__2                        NaN              MiSeq   \n",
       "fur__wt_dpd__1                    0.000000              MiSeq   \n",
       "fur__wt_dpd__2                    0.000000              MiSeq   \n",
       "fur__wt_fe__1                     1.060606              MiSeq   \n",
       "...                                    ...                ...   \n",
       "rpoB__rpoBE546V_glc__1            0.850000              MiSeq   \n",
       "rpoB__rpoBE672K_glc__1            0.880000              MiSeq   \n",
       "rpoB__rpoBE672K_glc__2            0.880000              MiSeq   \n",
       "rpoB__rpoBE672K_lb__1             1.400000              MiSeq   \n",
       "rpoB__rpoBE672K_lb__2             1.400000              MiSeq   \n",
       "\n",
       "                       Additional Details  Biological Replicates  \\\n",
       "Sample ID                                                          \n",
       "base__wt_glc__1                       NaN                      2   \n",
       "base__wt_glc__2                       NaN                      2   \n",
       "fur__wt_dpd__1                        NaN                      2   \n",
       "fur__wt_dpd__2                        NaN                      2   \n",
       "fur__wt_fe__1                         NaN                      2   \n",
       "...                                   ...                    ...   \n",
       "rpoB__rpoBE546V_glc__1                NaN                      1   \n",
       "rpoB__rpoBE672K_glc__1                NaN                      2   \n",
       "rpoB__rpoBE672K_glc__2                NaN                      2   \n",
       "rpoB__rpoBE672K_lb__1                 NaN                      2   \n",
       "rpoB__rpoBE672K_lb__2                 NaN                      2   \n",
       "\n",
       "                                                       DOI       GEO  \n",
       "Sample ID                                                             \n",
       "base__wt_glc__1                     doi.org/10.1101/080929  GSE65643  \n",
       "base__wt_glc__2                     doi.org/10.1101/080929  GSE65643  \n",
       "fur__wt_dpd__1                  doi.org/10.1038/ncomms5910  GSE54900  \n",
       "fur__wt_dpd__2                  doi.org/10.1038/ncomms5910  GSE54900  \n",
       "fur__wt_fe__1                   doi.org/10.1038/ncomms5910  GSE54900  \n",
       "...                                                    ...       ...  \n",
       "rpoB__rpoBE546V_glc__1  doi.org/10.1016/j.cels.2016.04.003  GSE59377  \n",
       "rpoB__rpoBE672K_glc__1  doi.org/10.1016/j.cels.2016.04.003  GSE59377  \n",
       "rpoB__rpoBE672K_glc__2  doi.org/10.1016/j.cels.2016.04.003  GSE59377  \n",
       "rpoB__rpoBE672K_lb__1   doi.org/10.1016/j.cels.2016.04.003  GSE59377  \n",
       "rpoB__rpoBE672K_lb__2   doi.org/10.1016/j.cels.2016.04.003  GSE59377  \n",
       "\n",
       "[102 rows x 19 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_table"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Above is the sample_table, also called metadata, which describes each sample in the data. The rows here correspond to the columns in the dataset (both have a total of 102). For a description of what each column means, see the 'Summary' sheet in Homework_Data.xlsx. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_product</th>\n",
       "      <th>COG</th>\n",
       "      <th>regulator</th>\n",
       "      <th>operon</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>b0002</th>\n",
       "      <td>thrA</td>\n",
       "      <td>fused aspartate kinase/homoserine dehydrogenase 1</td>\n",
       "      <td>Amino acid transport and metabolism</td>\n",
       "      <td>rpoD,arcA,crp,fnr,lrp</td>\n",
       "      <td>thrLABC</td>\n",
       "      <td>337.0</td>\n",
       "      <td>2799.0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0003</th>\n",
       "      <td>thrB</td>\n",
       "      <td>homoserine kinase</td>\n",
       "      <td>Nucleotide transport and metabolism</td>\n",
       "      <td>rpoD,arcA,crp,fnr,lrp</td>\n",
       "      <td>thrLABC</td>\n",
       "      <td>2801.0</td>\n",
       "      <td>3733.0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0004</th>\n",
       "      <td>thrC</td>\n",
       "      <td>threonine synthase</td>\n",
       "      <td>Amino acid transport and metabolism</td>\n",
       "      <td>rpoD,arcA,crp,fnr,lrp</td>\n",
       "      <td>thrLABC</td>\n",
       "      <td>3734.0</td>\n",
       "      <td>5020.0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0005</th>\n",
       "      <td>yaaX</td>\n",
       "      <td>DUF2502 domain-containing protein YaaX</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>yaaX</td>\n",
       "      <td>5234.0</td>\n",
       "      <td>5530.0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0006</th>\n",
       "      <td>yaaA</td>\n",
       "      <td>peroxide stress resistance protein YaaA</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>yaaA</td>\n",
       "      <td>5683.0</td>\n",
       "      <td>6459.0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b4688</th>\n",
       "      <td>ykgS</td>\n",
       "      <td>CP4-6 prophage%3B protein YkgS</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ykgS</td>\n",
       "      <td>290510.0</td>\n",
       "      <td>290638.0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b4693</th>\n",
       "      <td>aaaE</td>\n",
       "      <td>phage terminase</td>\n",
       "      <td>Transcription</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ymfTLMNR-beeE-jayE-ymfQ-ycfK-tfaP</td>\n",
       "      <td>1204401.0</td>\n",
       "      <td>1205537.0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b4696_1</th>\n",
       "      <td>ydeK</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>No COG Annotation</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ydeK</td>\n",
       "      <td>1594126.0</td>\n",
       "      <td>1597987.0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b4696_2</th>\n",
       "      <td>yneO_2</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>No COG Annotation</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ydeK</td>\n",
       "      <td>1592664.0</td>\n",
       "      <td>1594125.0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b4705</th>\n",
       "      <td>mntS</td>\n",
       "      <td>small protein MntS</td>\n",
       "      <td>No COG annotation</td>\n",
       "      <td>rpoD,mntR,nac</td>\n",
       "      <td>rybA-mntS</td>\n",
       "      <td>852869.0</td>\n",
       "      <td>852997.0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3887 rows Ã— 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        gene_name                                       gene_product  \\\n",
       "b0002        thrA  fused aspartate kinase/homoserine dehydrogenase 1   \n",
       "b0003        thrB                                  homoserine kinase   \n",
       "b0004        thrC                                 threonine synthase   \n",
       "b0005        yaaX             DUF2502 domain-containing protein YaaX   \n",
       "b0006        yaaA            peroxide stress resistance protein YaaA   \n",
       "...           ...                                                ...   \n",
       "b4688        ykgS                     CP4-6 prophage%3B protein YkgS   \n",
       "b4693        aaaE                                    phage terminase   \n",
       "b4696_1      ydeK                                            Unknown   \n",
       "b4696_2    yneO_2                                            Unknown   \n",
       "b4705        mntS                                 small protein MntS   \n",
       "\n",
       "                                         COG              regulator  \\\n",
       "b0002    Amino acid transport and metabolism  rpoD,arcA,crp,fnr,lrp   \n",
       "b0003    Nucleotide transport and metabolism  rpoD,arcA,crp,fnr,lrp   \n",
       "b0004    Amino acid transport and metabolism  rpoD,arcA,crp,fnr,lrp   \n",
       "b0005                       Function unknown                    NaN   \n",
       "b0006                       Function unknown                    NaN   \n",
       "...                                      ...                    ...   \n",
       "b4688                       Function unknown                    NaN   \n",
       "b4693                          Transcription                    NaN   \n",
       "b4696_1                    No COG Annotation                    NaN   \n",
       "b4696_2                    No COG Annotation                    NaN   \n",
       "b4705                      No COG annotation          rpoD,mntR,nac   \n",
       "\n",
       "                                    operon      start        end strand  \n",
       "b0002                              thrLABC      337.0     2799.0      +  \n",
       "b0003                              thrLABC     2801.0     3733.0      +  \n",
       "b0004                              thrLABC     3734.0     5020.0      +  \n",
       "b0005                                 yaaX     5234.0     5530.0      +  \n",
       "b0006                                 yaaA     5683.0     6459.0      -  \n",
       "...                                    ...        ...        ...    ...  \n",
       "b4688                                 ykgS   290510.0   290638.0      +  \n",
       "b4693    ymfTLMNR-beeE-jayE-ymfQ-ycfK-tfaP  1204401.0  1205537.0      +  \n",
       "b4696_1                               ydeK  1594126.0  1597987.0      -  \n",
       "b4696_2                               ydeK  1592664.0  1594125.0      -  \n",
       "b4705                            rybA-mntS   852869.0   852997.0      -  \n",
       "\n",
       "[3887 rows x 8 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_table"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Finally, this is the gene_table, which gives more information about each gene in the rows of the dataset. It is useful for gaining some quick information about any gene, or grouping genes together (often by COG). If you need more details on a gene, you should look it up in a database like [EcoCyc](https://ecocyc.org/). I recommend creating a free account there for this class."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Basic Data Access\n",
    "\n",
    "One of the most important skills you will develop in this class is the ability to easily navigate datasets in Pandas. This is surprisingly tricky, so it has been broken down very slowly in this section. [Here is a useful guide.](https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html)\n",
    "\n",
    "#### Columns\n",
    "\n",
    "Pandas DataFrames are essentially stored as a list of columns, which means that you don't need any fancy member functions to access them. If there are no spaces in the column name and it doesn't conflict with any pandas member function names, then you can access a column with a simple dot. For example, if you want to know the expression value for all genes in the base wild type glucose condition replicate 1, you access its column like this:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "b0002      7.629317\n",
       "b0003      7.219668\n",
       "b0004      7.343891\n",
       "b0005      4.378073\n",
       "b0006      4.793733\n",
       "             ...   \n",
       "b4688      1.714338\n",
       "b4693      0.602324\n",
       "b4696_1    1.834487\n",
       "b4696_2    2.138244\n",
       "b4705      4.292478\n",
       "Name: base__wt_glc__1, Length: 3887, dtype: float64"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data.base__wt_glc__1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If the column has a space or its name conflicts with a function name (e.g. \"sum\"), you have to use bracket notation. For example, there are spaces in the sample_table columns, so we would use this notation to see the carbon sources for all samples:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Sample ID\n",
       "base__wt_glc__1           glucose(2)\n",
       "base__wt_glc__2           glucose(2)\n",
       "fur__wt_dpd__1            glucose(2)\n",
       "fur__wt_dpd__2            glucose(2)\n",
       "fur__wt_fe__1             glucose(2)\n",
       "                             ...    \n",
       "rpoB__rpoBE546V_glc__1    glucose(4)\n",
       "rpoB__rpoBE672K_glc__1    glucose(4)\n",
       "rpoB__rpoBE672K_glc__2    glucose(4)\n",
       "rpoB__rpoBE672K_lb__1            NaN\n",
       "rpoB__rpoBE672K_lb__2            NaN\n",
       "Name: Carbon Source (g/L), Length: 102, dtype: object"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_table['Carbon Source (g/L)']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You can also use bracket notation with a list to slice out multiple columns at once. Let's say you wanted all gene expression values, but only for the two base conditions:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>base__wt_glc__1</th>\n",
       "      <th>base__wt_glc__2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>b0002</th>\n",
       "      <td>7.629317</td>\n",
       "      <td>7.714938</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0003</th>\n",
       "      <td>7.219668</td>\n",
       "      <td>7.294158</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0004</th>\n",
       "      <td>7.343891</td>\n",
       "      <td>7.434121</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0005</th>\n",
       "      <td>4.378073</td>\n",
       "      <td>4.338134</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0006</th>\n",
       "      <td>4.793733</td>\n",
       "      <td>4.781124</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b4688</th>\n",
       "      <td>1.714338</td>\n",
       "      <td>1.731933</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b4693</th>\n",
       "      <td>0.602324</td>\n",
       "      <td>0.988983</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b4696_1</th>\n",
       "      <td>1.834487</td>\n",
       "      <td>1.764432</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b4696_2</th>\n",
       "      <td>2.138244</td>\n",
       "      <td>2.182114</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b4705</th>\n",
       "      <td>4.292478</td>\n",
       "      <td>3.288338</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3887 rows Ã— 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         base__wt_glc__1  base__wt_glc__2\n",
       "b0002           7.629317         7.714938\n",
       "b0003           7.219668         7.294158\n",
       "b0004           7.343891         7.434121\n",
       "b0005           4.378073         4.338134\n",
       "b0006           4.793733         4.781124\n",
       "...                  ...              ...\n",
       "b4688           1.714338         1.731933\n",
       "b4693           0.602324         0.988983\n",
       "b4696_1         1.834487         1.764432\n",
       "b4696_2         2.138244         2.182114\n",
       "b4705           4.292478         3.288338\n",
       "\n",
       "[3887 rows x 2 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# make your list\n",
    "cols_of_interest = ['base__wt_glc__1', 'base__wt_glc__2']\n",
    "\n",
    "# index into the dataframe\n",
    "data[cols_of_interest]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Rows\n",
    "\n",
    "To get a row, you should use the \"loc\" function. Loc has many uses, which can get quite confusing. First, note that it can take a single row index or a list of row indices, just like the bracket notation for columns.\n",
    "\n",
    "In the first cell, we'll get the expression of the gene 'b0002' across all samples.\n",
    "\n",
    "In the second cell, we'll get the expression of three genes across all samples."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "base__wt_glc__1           7.629317\n",
       "base__wt_glc__2           7.714938\n",
       "fur__wt_dpd__1            8.113347\n",
       "fur__wt_dpd__2            8.240386\n",
       "fur__wt_fe__1             7.669640\n",
       "                            ...   \n",
       "rpoB__rpoBE546V_glc__1    7.574388\n",
       "rpoB__rpoBE672K_glc__1    7.625299\n",
       "rpoB__rpoBE672K_glc__2    7.624386\n",
       "rpoB__rpoBE672K_lb__1     5.640107\n",
       "rpoB__rpoBE672K_lb__2     5.652065\n",
       "Name: b0002, Length: 102, dtype: float64"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data.loc['b0002']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>base__wt_glc__1</th>\n",
       "      <th>base__wt_glc__2</th>\n",
       "      <th>fur__wt_dpd__1</th>\n",
       "      <th>fur__wt_dpd__2</th>\n",
       "      <th>fur__wt_fe__1</th>\n",
       "      <th>fur__wt_fe__2</th>\n",
       "      <th>fur__delfur_dpd__1</th>\n",
       "      <th>fur__delfur_dpd__2</th>\n",
       "      <th>fur__delfur_fe2__1</th>\n",
       "      <th>fur__delfur_fe2__2</th>\n",
       "      <th>...</th>\n",
       "      <th>rpoB__wt_lb__2</th>\n",
       "      <th>rpoB__rpoBE546V_031__1</th>\n",
       "      <th>rpoB__rpoBE546V_044__1</th>\n",
       "      <th>rpoB__rpoBE546V_lb__1</th>\n",
       "      <th>rpoB__rpoBE546V_lb__2</th>\n",
       "      <th>rpoB__rpoBE546V_glc__1</th>\n",
       "      <th>rpoB__rpoBE672K_glc__1</th>\n",
       "      <th>rpoB__rpoBE672K_glc__2</th>\n",
       "      <th>rpoB__rpoBE672K_lb__1</th>\n",
       "      <th>rpoB__rpoBE672K_lb__2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>b0002</th>\n",
       "      <td>7.629317</td>\n",
       "      <td>7.714938</td>\n",
       "      <td>8.113347</td>\n",
       "      <td>8.240386</td>\n",
       "      <td>7.669640</td>\n",
       "      <td>7.471594</td>\n",
       "      <td>6.915225</td>\n",
       "      <td>7.133376</td>\n",
       "      <td>7.784013</td>\n",
       "      <td>7.773122</td>\n",
       "      <td>...</td>\n",
       "      <td>5.191566</td>\n",
       "      <td>7.061956</td>\n",
       "      <td>7.142701</td>\n",
       "      <td>5.668539</td>\n",
       "      <td>5.649356</td>\n",
       "      <td>7.574388</td>\n",
       "      <td>7.625299</td>\n",
       "      <td>7.624386</td>\n",
       "      <td>5.640107</td>\n",
       "      <td>5.652065</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0003</th>\n",
       "      <td>7.219668</td>\n",
       "      <td>7.294158</td>\n",
       "      <td>7.918491</td>\n",
       "      <td>8.181859</td>\n",
       "      <td>7.470134</td>\n",
       "      <td>7.345953</td>\n",
       "      <td>6.652128</td>\n",
       "      <td>7.064316</td>\n",
       "      <td>7.554007</td>\n",
       "      <td>7.528505</td>\n",
       "      <td>...</td>\n",
       "      <td>5.547168</td>\n",
       "      <td>6.749462</td>\n",
       "      <td>6.945277</td>\n",
       "      <td>5.620395</td>\n",
       "      <td>5.646778</td>\n",
       "      <td>7.134694</td>\n",
       "      <td>7.374954</td>\n",
       "      <td>7.374000</td>\n",
       "      <td>5.369193</td>\n",
       "      <td>5.318577</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0004</th>\n",
       "      <td>7.343891</td>\n",
       "      <td>7.434121</td>\n",
       "      <td>7.248520</td>\n",
       "      <td>7.471647</td>\n",
       "      <td>7.205343</td>\n",
       "      <td>7.010567</td>\n",
       "      <td>6.059338</td>\n",
       "      <td>6.295427</td>\n",
       "      <td>7.059941</td>\n",
       "      <td>7.045726</td>\n",
       "      <td>...</td>\n",
       "      <td>5.549536</td>\n",
       "      <td>6.873381</td>\n",
       "      <td>6.964728</td>\n",
       "      <td>5.568881</td>\n",
       "      <td>5.612399</td>\n",
       "      <td>7.102728</td>\n",
       "      <td>7.538778</td>\n",
       "      <td>7.542841</td>\n",
       "      <td>5.402317</td>\n",
       "      <td>5.422052</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows Ã— 102 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       base__wt_glc__1  base__wt_glc__2  fur__wt_dpd__1  fur__wt_dpd__2  \\\n",
       "b0002         7.629317         7.714938        8.113347        8.240386   \n",
       "b0003         7.219668         7.294158        7.918491        8.181859   \n",
       "b0004         7.343891         7.434121        7.248520        7.471647   \n",
       "\n",
       "       fur__wt_fe__1  fur__wt_fe__2  fur__delfur_dpd__1  fur__delfur_dpd__2  \\\n",
       "b0002       7.669640       7.471594            6.915225            7.133376   \n",
       "b0003       7.470134       7.345953            6.652128            7.064316   \n",
       "b0004       7.205343       7.010567            6.059338            6.295427   \n",
       "\n",
       "       fur__delfur_fe2__1  fur__delfur_fe2__2  ...  rpoB__wt_lb__2  \\\n",
       "b0002            7.784013            7.773122  ...        5.191566   \n",
       "b0003            7.554007            7.528505  ...        5.547168   \n",
       "b0004            7.059941            7.045726  ...        5.549536   \n",
       "\n",
       "       rpoB__rpoBE546V_031__1  rpoB__rpoBE546V_044__1  rpoB__rpoBE546V_lb__1  \\\n",
       "b0002                7.061956                7.142701               5.668539   \n",
       "b0003                6.749462                6.945277               5.620395   \n",
       "b0004                6.873381                6.964728               5.568881   \n",
       "\n",
       "       rpoB__rpoBE546V_lb__2  rpoB__rpoBE546V_glc__1  rpoB__rpoBE672K_glc__1  \\\n",
       "b0002               5.649356                7.574388                7.625299   \n",
       "b0003               5.646778                7.134694                7.374954   \n",
       "b0004               5.612399                7.102728                7.538778   \n",
       "\n",
       "       rpoB__rpoBE672K_glc__2  rpoB__rpoBE672K_lb__1  rpoB__rpoBE672K_lb__2  \n",
       "b0002                7.624386               5.640107               5.652065  \n",
       "b0003                7.374000               5.369193               5.318577  \n",
       "b0004                7.542841               5.402317               5.422052  \n",
       "\n",
       "[3 rows x 102 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rows_of_interest = ['b0002', 'b0003', 'b0004']\n",
    "data.loc[rows_of_interest]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Single Values and Slicing Both Ways at Once\n",
    "\n",
    "To get a single value, you can either:\n",
    "\n",
    "- Access the column first, then use bracket notation\n",
    "- Use loc with a comma\n",
    "\n",
    "Which one you use is a personal preference, but I tend to use the first where possible because it requires a bit less typing. As a beginner, it might be easier to stick with loc. \n",
    "\n",
    "Let's say you want to know how much the gene b0004 is expressed in the base condition, replicate 1:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "7.62931651461589"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data.base__wt_glc__1['b0002']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "7.62931651461589"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data.loc['b0002', 'base__wt_glc__1'] # row, column"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's say you have a more specific slice you want. You can pass lists to both parts of loc if desired:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>base__wt_glc__1</th>\n",
       "      <th>base__wt_glc__2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>b0002</th>\n",
       "      <td>7.629317</td>\n",
       "      <td>7.714938</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0003</th>\n",
       "      <td>7.219668</td>\n",
       "      <td>7.294158</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0004</th>\n",
       "      <td>7.343891</td>\n",
       "      <td>7.434121</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       base__wt_glc__1  base__wt_glc__2\n",
       "b0002         7.629317         7.714938\n",
       "b0003         7.219668         7.294158\n",
       "b0004         7.343891         7.434121"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data.loc[rows_of_interest, cols_of_interest]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Other Ways of Indexing\n",
    "\n",
    "You may want to be familiar with [iloc](https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#indexing-integer), at, and iat. iloc, for example, allows you to index by number (as you would for lists and numpy arrays) instead of index names."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Data Access: Using the Sample Table\n",
    "\n",
    "Let's explore the information available to us in the sample table (for practice, you can try something similar with the gene table).\n",
    "\n",
    "First, we can list the columns easily to know which features are known to us:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Study', 'Condition ID', 'Replicate #', 'Strain Description',\n",
       "       'Base Media', 'Carbon Source (g/L)', 'Nitrogen Source (g/L)',\n",
       "       'Electron Acceptor', 'Trace Element Mixture', 'Supplement',\n",
       "       'Temperature (C)', 'pH', 'Culture Type', 'Growth Rate (1/hr)',\n",
       "       'Sequencing Machine', 'Additional Details', 'Biological Replicates',\n",
       "       'DOI', 'GEO'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_table.columns"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Unique()\n",
    "\n",
    "We can think of many of these columns as categorical variables, which have a finite set of available options. Let's use the [unique()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.unique.html) function to see which categories for Carbon Source exist in the dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['glucose(2)', 'glucose(4)', 'acetate(3.5)', 'fumarate(2.8)',\n",
       "       'glucose(5)', 'pyruvate(3.3)', 'acetate(2)', 'fructose(2)', nan],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_table['Carbon Source (g/L)'].unique()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "So, each sample's Carbon Source may be one of: three concentrations of glucose (2, 4, and 5 g/L), two concentrations of acetate (2 and 3.5 g/L), or fumarate, pyruvate, or fructose. The last entry in the list, \"nan\", is short for \"not a number\", which is what Pandas uses to fill empty cells. In this case, it indicates that no carbon source was added to the base media. By the way, you can replace all nan's with another value using [fillna()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.fillna.html?).\n",
    "\n",
    "The unique function is useful for knowing which options are available, and it can also be used to iterate through all options, as we will see in later sections.\n",
    "\n",
    "You may also want to know how many samples fall into each category; this can be done easily with the [value_counts()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.value_counts.html?highlight=value_counts#pandas.DataFrame.value_counts) function:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "glucose(2)       45\n",
       "glucose(4)       35\n",
       "acetate(2)        4\n",
       "fructose(2)       4\n",
       "acetate(3.5)      2\n",
       "pyruvate(3.3)     2\n",
       "glucose(5)        2\n",
       "fumarate(2.8)     2\n",
       "Name: Carbon Source (g/L), dtype: int64"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_table['Carbon Source (g/L)'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Boolean Series\n",
    "\n",
    "As shown above, there are 4 samples used fructose as a carbon source. Which samples are they, specifically? \n",
    "\n",
    "To find out, let's use the '==' operator to create a boolean (True/False) series for each sample. Note that the first equals sign is to assign this series to a variable, and the pair of equals signs compares each value in the column on the left to the value on the right."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Sample ID\n",
       "base__wt_glc__1           False\n",
       "base__wt_glc__2           False\n",
       "fur__wt_dpd__1            False\n",
       "fur__wt_dpd__2            False\n",
       "fur__wt_fe__1             False\n",
       "                          ...  \n",
       "rpoB__rpoBE546V_glc__1    False\n",
       "rpoB__rpoBE672K_glc__1    False\n",
       "rpoB__rpoBE672K_glc__2    False\n",
       "rpoB__rpoBE672K_lb__1     False\n",
       "rpoB__rpoBE672K_lb__2     False\n",
       "Name: Carbon Source (g/L), Length: 102, dtype: bool"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "uses_fructose = sample_table['Carbon Source (g/L)'] == 'fructose(2)'\n",
    "uses_fructose"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this series, each sample name is associated with a true/false value that indicates whether or not it uses fructose. We can use a series like this as a way of indexing! To get the names of the fructose-using samples, let's index into the row names themselves:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['cra_crp__wt_fru__1', 'cra_crp__wt_fru__2', 'cra_crp__delcra_fru__1',\n",
       "       'cra_crp__delcra_fru__2'],\n",
       "      dtype='object', name='Sample ID')"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fructose_samps = sample_table.index[uses_fructose]\n",
    "fructose_samps"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can also use the boolean series with \"loc\" to get the rest of the sample table details about these samples:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Study</th>\n",
       "      <th>Condition ID</th>\n",
       "      <th>Replicate #</th>\n",
       "      <th>Strain Description</th>\n",
       "      <th>Base Media</th>\n",
       "      <th>Carbon Source (g/L)</th>\n",
       "      <th>Nitrogen Source (g/L)</th>\n",
       "      <th>Electron Acceptor</th>\n",
       "      <th>Trace Element Mixture</th>\n",
       "      <th>Supplement</th>\n",
       "      <th>Temperature (C)</th>\n",
       "      <th>pH</th>\n",
       "      <th>Culture Type</th>\n",
       "      <th>Growth Rate (1/hr)</th>\n",
       "      <th>Sequencing Machine</th>\n",
       "      <th>Additional Details</th>\n",
       "      <th>Biological Replicates</th>\n",
       "      <th>DOI</th>\n",
       "      <th>GEO</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>cra_crp__wt_fru__1</th>\n",
       "      <td>Cra/Crp</td>\n",
       "      <td>wt_fru</td>\n",
       "      <td>1</td>\n",
       "      <td>Escherichia coli K-12 MG1655</td>\n",
       "      <td>M9</td>\n",
       "      <td>fructose(2)</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>O2</td>\n",
       "      <td>aebersold trace element mixture</td>\n",
       "      <td>NaN</td>\n",
       "      <td>37</td>\n",
       "      <td>7.0</td>\n",
       "      <td>Batch</td>\n",
       "      <td>0.53</td>\n",
       "      <td>MiSeq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>doi.org/10.1101/080929</td>\n",
       "      <td>GSE65643</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cra_crp__wt_fru__2</th>\n",
       "      <td>Cra/Crp</td>\n",
       "      <td>wt_fru</td>\n",
       "      <td>2</td>\n",
       "      <td>Escherichia coli K-12 MG1655</td>\n",
       "      <td>M9</td>\n",
       "      <td>fructose(2)</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>O2</td>\n",
       "      <td>aebersold trace element mixture</td>\n",
       "      <td>NaN</td>\n",
       "      <td>37</td>\n",
       "      <td>7.0</td>\n",
       "      <td>Batch</td>\n",
       "      <td>0.53</td>\n",
       "      <td>MiSeq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>doi.org/10.1101/080929</td>\n",
       "      <td>GSE65643</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cra_crp__delcra_fru__1</th>\n",
       "      <td>Cra/Crp</td>\n",
       "      <td>delcra_fru</td>\n",
       "      <td>1</td>\n",
       "      <td>Escherichia coli K-12 MG1655 del_cra</td>\n",
       "      <td>M9</td>\n",
       "      <td>fructose(2)</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>O2</td>\n",
       "      <td>aebersold trace element mixture</td>\n",
       "      <td>NaN</td>\n",
       "      <td>37</td>\n",
       "      <td>7.0</td>\n",
       "      <td>Batch</td>\n",
       "      <td>0.69</td>\n",
       "      <td>MiSeq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>doi.org/10.1101/080929</td>\n",
       "      <td>GSE65643</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cra_crp__delcra_fru__2</th>\n",
       "      <td>Cra/Crp</td>\n",
       "      <td>delcra_fru</td>\n",
       "      <td>2</td>\n",
       "      <td>Escherichia coli K-12 MG1655 del_cra</td>\n",
       "      <td>M9</td>\n",
       "      <td>fructose(2)</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>O2</td>\n",
       "      <td>aebersold trace element mixture</td>\n",
       "      <td>NaN</td>\n",
       "      <td>37</td>\n",
       "      <td>7.0</td>\n",
       "      <td>Batch</td>\n",
       "      <td>0.69</td>\n",
       "      <td>MiSeq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>doi.org/10.1101/080929</td>\n",
       "      <td>GSE65643</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          Study Condition ID  Replicate #  \\\n",
       "Sample ID                                                   \n",
       "cra_crp__wt_fru__1      Cra/Crp       wt_fru            1   \n",
       "cra_crp__wt_fru__2      Cra/Crp       wt_fru            2   \n",
       "cra_crp__delcra_fru__1  Cra/Crp   delcra_fru            1   \n",
       "cra_crp__delcra_fru__2  Cra/Crp   delcra_fru            2   \n",
       "\n",
       "                                          Strain Description Base Media  \\\n",
       "Sample ID                                                                 \n",
       "cra_crp__wt_fru__1              Escherichia coli K-12 MG1655         M9   \n",
       "cra_crp__wt_fru__2              Escherichia coli K-12 MG1655         M9   \n",
       "cra_crp__delcra_fru__1  Escherichia coli K-12 MG1655 del_cra         M9   \n",
       "cra_crp__delcra_fru__2  Escherichia coli K-12 MG1655 del_cra         M9   \n",
       "\n",
       "                       Carbon Source (g/L) Nitrogen Source (g/L)  \\\n",
       "Sample ID                                                          \n",
       "cra_crp__wt_fru__1             fructose(2)              NH4Cl(1)   \n",
       "cra_crp__wt_fru__2             fructose(2)              NH4Cl(1)   \n",
       "cra_crp__delcra_fru__1         fructose(2)              NH4Cl(1)   \n",
       "cra_crp__delcra_fru__2         fructose(2)              NH4Cl(1)   \n",
       "\n",
       "                       Electron Acceptor            Trace Element Mixture  \\\n",
       "Sample ID                                                                   \n",
       "cra_crp__wt_fru__1                    O2  aebersold trace element mixture   \n",
       "cra_crp__wt_fru__2                    O2  aebersold trace element mixture   \n",
       "cra_crp__delcra_fru__1                O2  aebersold trace element mixture   \n",
       "cra_crp__delcra_fru__2                O2  aebersold trace element mixture   \n",
       "\n",
       "                       Supplement  Temperature (C)   pH Culture Type  \\\n",
       "Sample ID                                                              \n",
       "cra_crp__wt_fru__1            NaN               37  7.0        Batch   \n",
       "cra_crp__wt_fru__2            NaN               37  7.0        Batch   \n",
       "cra_crp__delcra_fru__1        NaN               37  7.0        Batch   \n",
       "cra_crp__delcra_fru__2        NaN               37  7.0        Batch   \n",
       "\n",
       "                        Growth Rate (1/hr) Sequencing Machine  \\\n",
       "Sample ID                                                       \n",
       "cra_crp__wt_fru__1                    0.53              MiSeq   \n",
       "cra_crp__wt_fru__2                    0.53              MiSeq   \n",
       "cra_crp__delcra_fru__1                0.69              MiSeq   \n",
       "cra_crp__delcra_fru__2                0.69              MiSeq   \n",
       "\n",
       "                       Additional Details  Biological Replicates  \\\n",
       "Sample ID                                                          \n",
       "cra_crp__wt_fru__1                    NaN                      2   \n",
       "cra_crp__wt_fru__2                    NaN                      2   \n",
       "cra_crp__delcra_fru__1                NaN                      2   \n",
       "cra_crp__delcra_fru__2                NaN                      2   \n",
       "\n",
       "                                           DOI       GEO  \n",
       "Sample ID                                                 \n",
       "cra_crp__wt_fru__1      doi.org/10.1101/080929  GSE65643  \n",
       "cra_crp__wt_fru__2      doi.org/10.1101/080929  GSE65643  \n",
       "cra_crp__delcra_fru__1  doi.org/10.1101/080929  GSE65643  \n",
       "cra_crp__delcra_fru__2  doi.org/10.1101/080929  GSE65643  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_table.loc[uses_fructose]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Lastly, let's say we want to get the gene expression data relating to these samples. In the data variable, samples are *columns*, even though they are rows in the sample table. The safest way to do this is to use the index names, instead of the boolean series."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cra_crp__wt_fru__1</th>\n",
       "      <th>cra_crp__wt_fru__2</th>\n",
       "      <th>cra_crp__delcra_fru__1</th>\n",
       "      <th>cra_crp__delcra_fru__2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>b0002</th>\n",
       "      <td>7.835301</td>\n",
       "      <td>7.688688</td>\n",
       "      <td>8.108317</td>\n",
       "      <td>8.041206</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0003</th>\n",
       "      <td>7.014402</td>\n",
       "      <td>7.101097</td>\n",
       "      <td>7.900230</td>\n",
       "      <td>7.893612</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0004</th>\n",
       "      <td>7.016143</td>\n",
       "      <td>6.899328</td>\n",
       "      <td>7.471183</td>\n",
       "      <td>7.386816</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0005</th>\n",
       "      <td>4.128200</td>\n",
       "      <td>4.271289</td>\n",
       "      <td>3.660925</td>\n",
       "      <td>3.924343</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0006</th>\n",
       "      <td>4.472146</td>\n",
       "      <td>4.615870</td>\n",
       "      <td>4.846699</td>\n",
       "      <td>4.821392</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b4688</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.268005</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b4693</th>\n",
       "      <td>1.839114</td>\n",
       "      <td>1.445625</td>\n",
       "      <td>1.872467</td>\n",
       "      <td>2.028900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b4696_1</th>\n",
       "      <td>1.581827</td>\n",
       "      <td>1.552505</td>\n",
       "      <td>2.136073</td>\n",
       "      <td>2.242039</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b4696_2</th>\n",
       "      <td>1.745700</td>\n",
       "      <td>1.945509</td>\n",
       "      <td>2.472103</td>\n",
       "      <td>2.256234</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b4705</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.307742</td>\n",
       "      <td>1.046738</td>\n",
       "      <td>0.956291</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3887 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         cra_crp__wt_fru__1  cra_crp__wt_fru__2  cra_crp__delcra_fru__1  \\\n",
       "b0002              7.835301            7.688688                8.108317   \n",
       "b0003              7.014402            7.101097                7.900230   \n",
       "b0004              7.016143            6.899328                7.471183   \n",
       "b0005              4.128200            4.271289                3.660925   \n",
       "b0006              4.472146            4.615870                4.846699   \n",
       "...                     ...                 ...                     ...   \n",
       "b4688              0.000000            1.268005                0.000000   \n",
       "b4693              1.839114            1.445625                1.872467   \n",
       "b4696_1            1.581827            1.552505                2.136073   \n",
       "b4696_2            1.745700            1.945509                2.472103   \n",
       "b4705              0.000000            1.307742                1.046738   \n",
       "\n",
       "         cra_crp__delcra_fru__2  \n",
       "b0002                  8.041206  \n",
       "b0003                  7.893612  \n",
       "b0004                  7.386816  \n",
       "b0005                  3.924343  \n",
       "b0006                  4.821392  \n",
       "...                         ...  \n",
       "b4688                  0.000000  \n",
       "b4693                  2.028900  \n",
       "b4696_1                2.242039  \n",
       "b4696_2                2.256234  \n",
       "b4705                  0.956291  \n",
       "\n",
       "[3887 rows x 4 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fru_data = data[fructose_samps]\n",
    "fru_data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's repeat the above data access, but instead get the glucose(2) samples:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>base__wt_glc__1</th>\n",
       "      <th>base__wt_glc__2</th>\n",
       "      <th>fur__wt_dpd__1</th>\n",
       "      <th>fur__wt_dpd__2</th>\n",
       "      <th>fur__wt_fe__1</th>\n",
       "      <th>fur__wt_fe__2</th>\n",
       "      <th>fur__delfur_dpd__1</th>\n",
       "      <th>fur__delfur_dpd__2</th>\n",
       "      <th>fur__delfur_fe2__1</th>\n",
       "      <th>fur__delfur_fe2__2</th>\n",
       "      <th>...</th>\n",
       "      <th>nac_ntrc__bw_delnac_cytd__2</th>\n",
       "      <th>nac_ntrc__bw_delnac_gln__1</th>\n",
       "      <th>nac_ntrc__bw_delnac_gln__2</th>\n",
       "      <th>nac_ntrc__bw_delnac_nh4__1</th>\n",
       "      <th>nac_ntrc__bw_delnac_nh4__2</th>\n",
       "      <th>ompr__wt_nacl__1</th>\n",
       "      <th>ompr__bw_delompr_nacl__1</th>\n",
       "      <th>ompr__bw_delompr_nacl__2</th>\n",
       "      <th>cra_crp__delcra_glc__1</th>\n",
       "      <th>cra_crp__delcra_glc__2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>b0002</th>\n",
       "      <td>7.629317</td>\n",
       "      <td>7.714938</td>\n",
       "      <td>8.113347</td>\n",
       "      <td>8.240386</td>\n",
       "      <td>7.669640</td>\n",
       "      <td>7.471594</td>\n",
       "      <td>6.915225</td>\n",
       "      <td>7.133376</td>\n",
       "      <td>7.784013</td>\n",
       "      <td>7.773122</td>\n",
       "      <td>...</td>\n",
       "      <td>7.582810</td>\n",
       "      <td>7.139237</td>\n",
       "      <td>7.120299</td>\n",
       "      <td>7.982307</td>\n",
       "      <td>7.992733</td>\n",
       "      <td>5.927571</td>\n",
       "      <td>5.241910</td>\n",
       "      <td>5.251097</td>\n",
       "      <td>7.768612</td>\n",
       "      <td>7.800013</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0003</th>\n",
       "      <td>7.219668</td>\n",
       "      <td>7.294158</td>\n",
       "      <td>7.918491</td>\n",
       "      <td>8.181859</td>\n",
       "      <td>7.470134</td>\n",
       "      <td>7.345953</td>\n",
       "      <td>6.652128</td>\n",
       "      <td>7.064316</td>\n",
       "      <td>7.554007</td>\n",
       "      <td>7.528505</td>\n",
       "      <td>...</td>\n",
       "      <td>7.235787</td>\n",
       "      <td>6.801614</td>\n",
       "      <td>6.784171</td>\n",
       "      <td>7.594932</td>\n",
       "      <td>7.649324</td>\n",
       "      <td>6.059250</td>\n",
       "      <td>5.396705</td>\n",
       "      <td>5.435278</td>\n",
       "      <td>7.721704</td>\n",
       "      <td>7.674799</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0004</th>\n",
       "      <td>7.343891</td>\n",
       "      <td>7.434121</td>\n",
       "      <td>7.248520</td>\n",
       "      <td>7.471647</td>\n",
       "      <td>7.205343</td>\n",
       "      <td>7.010567</td>\n",
       "      <td>6.059338</td>\n",
       "      <td>6.295427</td>\n",
       "      <td>7.059941</td>\n",
       "      <td>7.045726</td>\n",
       "      <td>...</td>\n",
       "      <td>7.127754</td>\n",
       "      <td>6.788806</td>\n",
       "      <td>6.734994</td>\n",
       "      <td>7.630499</td>\n",
       "      <td>7.613657</td>\n",
       "      <td>5.391694</td>\n",
       "      <td>4.905536</td>\n",
       "      <td>4.847550</td>\n",
       "      <td>7.227646</td>\n",
       "      <td>7.310178</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0005</th>\n",
       "      <td>4.378073</td>\n",
       "      <td>4.338134</td>\n",
       "      <td>3.758424</td>\n",
       "      <td>3.698816</td>\n",
       "      <td>4.655322</td>\n",
       "      <td>4.443782</td>\n",
       "      <td>3.207263</td>\n",
       "      <td>3.296815</td>\n",
       "      <td>4.524755</td>\n",
       "      <td>4.253390</td>\n",
       "      <td>...</td>\n",
       "      <td>2.835923</td>\n",
       "      <td>2.888184</td>\n",
       "      <td>2.542888</td>\n",
       "      <td>4.085248</td>\n",
       "      <td>3.806096</td>\n",
       "      <td>2.152347</td>\n",
       "      <td>2.700117</td>\n",
       "      <td>2.871590</td>\n",
       "      <td>4.041135</td>\n",
       "      <td>4.025525</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0006</th>\n",
       "      <td>4.793733</td>\n",
       "      <td>4.781124</td>\n",
       "      <td>4.696611</td>\n",
       "      <td>4.701444</td>\n",
       "      <td>4.687029</td>\n",
       "      <td>4.724970</td>\n",
       "      <td>4.635023</td>\n",
       "      <td>4.755118</td>\n",
       "      <td>4.756608</td>\n",
       "      <td>4.725641</td>\n",
       "      <td>...</td>\n",
       "      <td>4.880251</td>\n",
       "      <td>4.944349</td>\n",
       "      <td>5.038410</td>\n",
       "      <td>4.912290</td>\n",
       "      <td>4.848897</td>\n",
       "      <td>4.238347</td>\n",
       "      <td>4.285009</td>\n",
       "      <td>4.360182</td>\n",
       "      <td>4.805900</td>\n",
       "      <td>4.811306</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b4688</th>\n",
       "      <td>1.714338</td>\n",
       "      <td>1.731933</td>\n",
       "      <td>0.553218</td>\n",
       "      <td>0.571096</td>\n",
       "      <td>1.680226</td>\n",
       "      <td>1.482264</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.514211</td>\n",
       "      <td>1.828054</td>\n",
       "      <td>1.863042</td>\n",
       "      <td>...</td>\n",
       "      <td>0.620424</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.230012</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.798580</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.618245</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b4693</th>\n",
       "      <td>0.602324</td>\n",
       "      <td>0.988983</td>\n",
       "      <td>1.739311</td>\n",
       "      <td>1.502915</td>\n",
       "      <td>1.217188</td>\n",
       "      <td>1.480901</td>\n",
       "      <td>2.569072</td>\n",
       "      <td>2.280928</td>\n",
       "      <td>2.958956</td>\n",
       "      <td>2.947872</td>\n",
       "      <td>...</td>\n",
       "      <td>1.377430</td>\n",
       "      <td>1.232259</td>\n",
       "      <td>1.014440</td>\n",
       "      <td>1.123002</td>\n",
       "      <td>1.628627</td>\n",
       "      <td>1.944512</td>\n",
       "      <td>1.079706</td>\n",
       "      <td>0.766141</td>\n",
       "      <td>1.693128</td>\n",
       "      <td>1.616979</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b4696_1</th>\n",
       "      <td>1.834487</td>\n",
       "      <td>1.764432</td>\n",
       "      <td>2.607990</td>\n",
       "      <td>2.522697</td>\n",
       "      <td>1.431101</td>\n",
       "      <td>1.396323</td>\n",
       "      <td>2.433068</td>\n",
       "      <td>2.306292</td>\n",
       "      <td>1.711284</td>\n",
       "      <td>1.685786</td>\n",
       "      <td>...</td>\n",
       "      <td>2.537078</td>\n",
       "      <td>2.537734</td>\n",
       "      <td>2.571499</td>\n",
       "      <td>1.813043</td>\n",
       "      <td>1.994967</td>\n",
       "      <td>2.523774</td>\n",
       "      <td>2.533935</td>\n",
       "      <td>2.260880</td>\n",
       "      <td>1.858779</td>\n",
       "      <td>1.923097</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b4696_2</th>\n",
       "      <td>2.138244</td>\n",
       "      <td>2.182114</td>\n",
       "      <td>2.704712</td>\n",
       "      <td>2.771110</td>\n",
       "      <td>1.829371</td>\n",
       "      <td>1.525206</td>\n",
       "      <td>2.740794</td>\n",
       "      <td>2.715689</td>\n",
       "      <td>2.110518</td>\n",
       "      <td>2.159843</td>\n",
       "      <td>...</td>\n",
       "      <td>2.537078</td>\n",
       "      <td>2.537734</td>\n",
       "      <td>2.571499</td>\n",
       "      <td>1.813043</td>\n",
       "      <td>1.994967</td>\n",
       "      <td>2.738769</td>\n",
       "      <td>2.533935</td>\n",
       "      <td>2.260880</td>\n",
       "      <td>2.082040</td>\n",
       "      <td>2.032237</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b4705</th>\n",
       "      <td>4.292478</td>\n",
       "      <td>3.288338</td>\n",
       "      <td>0.775085</td>\n",
       "      <td>0.797724</td>\n",
       "      <td>3.272305</td>\n",
       "      <td>2.919118</td>\n",
       "      <td>1.497251</td>\n",
       "      <td>0.725308</td>\n",
       "      <td>1.933021</td>\n",
       "      <td>2.492378</td>\n",
       "      <td>...</td>\n",
       "      <td>0.859645</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.803122</td>\n",
       "      <td>1.138273</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.699145</td>\n",
       "      <td>1.311520</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3887 rows Ã— 45 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         base__wt_glc__1  base__wt_glc__2  fur__wt_dpd__1  fur__wt_dpd__2  \\\n",
       "b0002           7.629317         7.714938        8.113347        8.240386   \n",
       "b0003           7.219668         7.294158        7.918491        8.181859   \n",
       "b0004           7.343891         7.434121        7.248520        7.471647   \n",
       "b0005           4.378073         4.338134        3.758424        3.698816   \n",
       "b0006           4.793733         4.781124        4.696611        4.701444   \n",
       "...                  ...              ...             ...             ...   \n",
       "b4688           1.714338         1.731933        0.553218        0.571096   \n",
       "b4693           0.602324         0.988983        1.739311        1.502915   \n",
       "b4696_1         1.834487         1.764432        2.607990        2.522697   \n",
       "b4696_2         2.138244         2.182114        2.704712        2.771110   \n",
       "b4705           4.292478         3.288338        0.775085        0.797724   \n",
       "\n",
       "         fur__wt_fe__1  fur__wt_fe__2  fur__delfur_dpd__1  fur__delfur_dpd__2  \\\n",
       "b0002         7.669640       7.471594            6.915225            7.133376   \n",
       "b0003         7.470134       7.345953            6.652128            7.064316   \n",
       "b0004         7.205343       7.010567            6.059338            6.295427   \n",
       "b0005         4.655322       4.443782            3.207263            3.296815   \n",
       "b0006         4.687029       4.724970            4.635023            4.755118   \n",
       "...                ...            ...                 ...                 ...   \n",
       "b4688         1.680226       1.482264            0.000000            0.514211   \n",
       "b4693         1.217188       1.480901            2.569072            2.280928   \n",
       "b4696_1       1.431101       1.396323            2.433068            2.306292   \n",
       "b4696_2       1.829371       1.525206            2.740794            2.715689   \n",
       "b4705         3.272305       2.919118            1.497251            0.725308   \n",
       "\n",
       "         fur__delfur_fe2__1  fur__delfur_fe2__2  ...  \\\n",
       "b0002              7.784013            7.773122  ...   \n",
       "b0003              7.554007            7.528505  ...   \n",
       "b0004              7.059941            7.045726  ...   \n",
       "b0005              4.524755            4.253390  ...   \n",
       "b0006              4.756608            4.725641  ...   \n",
       "...                     ...                 ...  ...   \n",
       "b4688              1.828054            1.863042  ...   \n",
       "b4693              2.958956            2.947872  ...   \n",
       "b4696_1            1.711284            1.685786  ...   \n",
       "b4696_2            2.110518            2.159843  ...   \n",
       "b4705              1.933021            2.492378  ...   \n",
       "\n",
       "         nac_ntrc__bw_delnac_cytd__2  nac_ntrc__bw_delnac_gln__1  \\\n",
       "b0002                       7.582810                    7.139237   \n",
       "b0003                       7.235787                    6.801614   \n",
       "b0004                       7.127754                    6.788806   \n",
       "b0005                       2.835923                    2.888184   \n",
       "b0006                       4.880251                    4.944349   \n",
       "...                              ...                         ...   \n",
       "b4688                       0.620424                    0.000000   \n",
       "b4693                       1.377430                    1.232259   \n",
       "b4696_1                     2.537078                    2.537734   \n",
       "b4696_2                     2.537078                    2.537734   \n",
       "b4705                       0.859645                    0.000000   \n",
       "\n",
       "         nac_ntrc__bw_delnac_gln__2  nac_ntrc__bw_delnac_nh4__1  \\\n",
       "b0002                      7.120299                    7.982307   \n",
       "b0003                      6.784171                    7.594932   \n",
       "b0004                      6.734994                    7.630499   \n",
       "b0005                      2.542888                    4.085248   \n",
       "b0006                      5.038410                    4.912290   \n",
       "...                             ...                         ...   \n",
       "b4688                      1.230012                    0.000000   \n",
       "b4693                      1.014440                    1.123002   \n",
       "b4696_1                    2.571499                    1.813043   \n",
       "b4696_2                    2.571499                    1.813043   \n",
       "b4705                      0.000000                    0.000000   \n",
       "\n",
       "         nac_ntrc__bw_delnac_nh4__2  ompr__wt_nacl__1  \\\n",
       "b0002                      7.992733          5.927571   \n",
       "b0003                      7.649324          6.059250   \n",
       "b0004                      7.613657          5.391694   \n",
       "b0005                      3.806096          2.152347   \n",
       "b0006                      4.848897          4.238347   \n",
       "...                             ...               ...   \n",
       "b4688                      0.000000          0.798580   \n",
       "b4693                      1.628627          1.944512   \n",
       "b4696_1                    1.994967          2.523774   \n",
       "b4696_2                    1.994967          2.738769   \n",
       "b4705                      0.000000          2.803122   \n",
       "\n",
       "         ompr__bw_delompr_nacl__1  ompr__bw_delompr_nacl__2  \\\n",
       "b0002                    5.241910                  5.251097   \n",
       "b0003                    5.396705                  5.435278   \n",
       "b0004                    4.905536                  4.847550   \n",
       "b0005                    2.700117                  2.871590   \n",
       "b0006                    4.285009                  4.360182   \n",
       "...                           ...                       ...   \n",
       "b4688                    0.000000                  0.000000   \n",
       "b4693                    1.079706                  0.766141   \n",
       "b4696_1                  2.533935                  2.260880   \n",
       "b4696_2                  2.533935                  2.260880   \n",
       "b4705                    1.138273                  0.000000   \n",
       "\n",
       "         cra_crp__delcra_glc__1  cra_crp__delcra_glc__2  \n",
       "b0002                  7.768612                7.800013  \n",
       "b0003                  7.721704                7.674799  \n",
       "b0004                  7.227646                7.310178  \n",
       "b0005                  4.041135                4.025525  \n",
       "b0006                  4.805900                4.811306  \n",
       "...                         ...                     ...  \n",
       "b4688                  0.000000                0.618245  \n",
       "b4693                  1.693128                1.616979  \n",
       "b4696_1                1.858779                1.923097  \n",
       "b4696_2                2.082040                2.032237  \n",
       "b4705                  1.699145                1.311520  \n",
       "\n",
       "[3887 rows x 45 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "uses_glu2 = sample_table['Carbon Source (g/L)'] == 'glucose(2)' # boolean series\n",
    "glu2_samps = sample_table.index[uses_glu2] # sample names\n",
    "glu2_data = data[glu2_samps] # slice of data\n",
    "glu2_data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Basic Statistics, Math, and Sorting: Glu v Fru Example\n",
    "\n",
    "How does the choice of glucose versus fructose affect gene expression across all relevant conditions in the dataset?\n",
    "\n",
    "We are already partway through answering this question, since we collected the data in section 5.1.2. Let's do a simple comparison of means - you'll explore more rigorous comparisons in the graded homeworks.\n",
    "\n",
    "#### Results DataFrame and DataFrame Means\n",
    "\n",
    "We'll start by computing the mean expression across *all columns* of both the fru_data and glu2_data. Think carefully about what this means: for each gene, we pool all conditions that have the given carbon source. \n",
    "\n",
    "To do this, use the [Pandas mean() function](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.mean.html). By default, this function takes the mean down each column, but we want to take the mean across each row, so we pass the argument \"axis = 1\" to switch directions.\n",
    "\n",
    "We'll save our results in a results DataFrame - this is a good way keep track of all your new computations in one place. Alternatively, you could have a bunch of separate Series. You wouldn't lose points for that approach, but it is a bit harder to keep it organized. \n",
    "\n",
    "Making the results DataFrame involves making a DataFrame nearly from scratch, which is covered in [this documentation](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.html). You don't need to populate it with values right away, but you should probably indicate what the indices and columns are from the get-go, as shown below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "b0002      7.918378\n",
       "b0003      7.477335\n",
       "b0004      7.193367\n",
       "b0005      3.996189\n",
       "b0006      4.689027\n",
       "             ...   \n",
       "b4688      0.317001\n",
       "b4693      1.796527\n",
       "b4696_1    1.878111\n",
       "b4696_2    2.104886\n",
       "b4705      0.827693\n",
       "Length: 3887, dtype: float64"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fru_data.mean(axis = 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>glu2_avg</th>\n",
       "      <th>fru_avg</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>b0002</th>\n",
       "      <td>7.102886</td>\n",
       "      <td>7.918378</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0003</th>\n",
       "      <td>6.957513</td>\n",
       "      <td>7.477335</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0004</th>\n",
       "      <td>6.640754</td>\n",
       "      <td>7.193367</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0005</th>\n",
       "      <td>3.403781</td>\n",
       "      <td>3.996189</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0006</th>\n",
       "      <td>4.814655</td>\n",
       "      <td>4.689027</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b4688</th>\n",
       "      <td>0.692460</td>\n",
       "      <td>0.317001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b4693</th>\n",
       "      <td>1.626249</td>\n",
       "      <td>1.796527</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b4696_1</th>\n",
       "      <td>2.232605</td>\n",
       "      <td>1.878111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b4696_2</th>\n",
       "      <td>2.400382</td>\n",
       "      <td>2.104886</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b4705</th>\n",
       "      <td>1.696125</td>\n",
       "      <td>0.827693</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3887 rows Ã— 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         glu2_avg   fru_avg\n",
       "b0002    7.102886  7.918378\n",
       "b0003    6.957513  7.477335\n",
       "b0004    6.640754  7.193367\n",
       "b0005    3.403781  3.996189\n",
       "b0006    4.814655  4.689027\n",
       "...           ...       ...\n",
       "b4688    0.692460  0.317001\n",
       "b4693    1.626249  1.796527\n",
       "b4696_1  2.232605  1.878111\n",
       "b4696_2  2.400382  2.104886\n",
       "b4705    1.696125  0.827693\n",
       "\n",
       "[3887 rows x 2 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# set up your results dataframe\n",
    "res_df = pd.DataFrame(index = data.index, # index is the same as data, i.e. one row for each gene\n",
    "                      columns = ['glu2_avg', 'fru_avg']) # columns are the results (carbon source averages)\n",
    "\n",
    "# fill in the glu2_avg column\n",
    "res_df['glu2_avg'] = glu2_data.mean(axis = 1)\n",
    "res_df['fru_avg'] = fru_data.mean(axis = 1)\n",
    "\n",
    "# display the results\n",
    "res_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Subtraction and Absolute Value\n",
    "\n",
    "As you can see above, some of the values are pretty close. It would be nice to know which genes are affected the most, which isn't clear yet. Let's get a metric for determining that - how about the difference between means?\n",
    "\n",
    "Let's subtract one of our columns from the other, and make the result a new column in the dataframe:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>glu2_avg</th>\n",
       "      <th>fru_avg</th>\n",
       "      <th>difference</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>b0002</th>\n",
       "      <td>7.102886</td>\n",
       "      <td>7.918378</td>\n",
       "      <td>0.815492</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0003</th>\n",
       "      <td>6.957513</td>\n",
       "      <td>7.477335</td>\n",
       "      <td>0.519822</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0004</th>\n",
       "      <td>6.640754</td>\n",
       "      <td>7.193367</td>\n",
       "      <td>0.552614</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0005</th>\n",
       "      <td>3.403781</td>\n",
       "      <td>3.996189</td>\n",
       "      <td>0.592408</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0006</th>\n",
       "      <td>4.814655</td>\n",
       "      <td>4.689027</td>\n",
       "      <td>-0.125628</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b4688</th>\n",
       "      <td>0.692460</td>\n",
       "      <td>0.317001</td>\n",
       "      <td>-0.375459</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b4693</th>\n",
       "      <td>1.626249</td>\n",
       "      <td>1.796527</td>\n",
       "      <td>0.170278</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b4696_1</th>\n",
       "      <td>2.232605</td>\n",
       "      <td>1.878111</td>\n",
       "      <td>-0.354493</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b4696_2</th>\n",
       "      <td>2.400382</td>\n",
       "      <td>2.104886</td>\n",
       "      <td>-0.295496</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b4705</th>\n",
       "      <td>1.696125</td>\n",
       "      <td>0.827693</td>\n",
       "      <td>-0.868432</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3887 rows Ã— 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         glu2_avg   fru_avg  difference\n",
       "b0002    7.102886  7.918378    0.815492\n",
       "b0003    6.957513  7.477335    0.519822\n",
       "b0004    6.640754  7.193367    0.552614\n",
       "b0005    3.403781  3.996189    0.592408\n",
       "b0006    4.814655  4.689027   -0.125628\n",
       "...           ...       ...         ...\n",
       "b4688    0.692460  0.317001   -0.375459\n",
       "b4693    1.626249  1.796527    0.170278\n",
       "b4696_1  2.232605  1.878111   -0.354493\n",
       "b4696_2  2.400382  2.104886   -0.295496\n",
       "b4705    1.696125  0.827693   -0.868432\n",
       "\n",
       "[3887 rows x 3 columns]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res_df['difference'] = res_df.fru_avg - res_df.glu2_avg\n",
    "\n",
    "res_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Since we want to know \"which genes are affected most,\" we don't care if the genes are more highly expressed in glucose or in fructose. Therefore, we should take the [absolute value](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.abs.html). Again, save it as a new column:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>glu2_avg</th>\n",
       "      <th>fru_avg</th>\n",
       "      <th>difference</th>\n",
       "      <th>abs_diff</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>b0002</th>\n",
       "      <td>7.102886</td>\n",
       "      <td>7.918378</td>\n",
       "      <td>0.815492</td>\n",
       "      <td>0.815492</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0003</th>\n",
       "      <td>6.957513</td>\n",
       "      <td>7.477335</td>\n",
       "      <td>0.519822</td>\n",
       "      <td>0.519822</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0004</th>\n",
       "      <td>6.640754</td>\n",
       "      <td>7.193367</td>\n",
       "      <td>0.552614</td>\n",
       "      <td>0.552614</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0005</th>\n",
       "      <td>3.403781</td>\n",
       "      <td>3.996189</td>\n",
       "      <td>0.592408</td>\n",
       "      <td>0.592408</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0006</th>\n",
       "      <td>4.814655</td>\n",
       "      <td>4.689027</td>\n",
       "      <td>-0.125628</td>\n",
       "      <td>0.125628</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b4688</th>\n",
       "      <td>0.692460</td>\n",
       "      <td>0.317001</td>\n",
       "      <td>-0.375459</td>\n",
       "      <td>0.375459</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b4693</th>\n",
       "      <td>1.626249</td>\n",
       "      <td>1.796527</td>\n",
       "      <td>0.170278</td>\n",
       "      <td>0.170278</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b4696_1</th>\n",
       "      <td>2.232605</td>\n",
       "      <td>1.878111</td>\n",
       "      <td>-0.354493</td>\n",
       "      <td>0.354493</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b4696_2</th>\n",
       "      <td>2.400382</td>\n",
       "      <td>2.104886</td>\n",
       "      <td>-0.295496</td>\n",
       "      <td>0.295496</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b4705</th>\n",
       "      <td>1.696125</td>\n",
       "      <td>0.827693</td>\n",
       "      <td>-0.868432</td>\n",
       "      <td>0.868432</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3887 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         glu2_avg   fru_avg  difference  abs_diff\n",
       "b0002    7.102886  7.918378    0.815492  0.815492\n",
       "b0003    6.957513  7.477335    0.519822  0.519822\n",
       "b0004    6.640754  7.193367    0.552614  0.552614\n",
       "b0005    3.403781  3.996189    0.592408  0.592408\n",
       "b0006    4.814655  4.689027   -0.125628  0.125628\n",
       "...           ...       ...         ...       ...\n",
       "b4688    0.692460  0.317001   -0.375459  0.375459\n",
       "b4693    1.626249  1.796527    0.170278  0.170278\n",
       "b4696_1  2.232605  1.878111   -0.354493  0.354493\n",
       "b4696_2  2.400382  2.104886   -0.295496  0.295496\n",
       "b4705    1.696125  0.827693   -0.868432  0.868432\n",
       "\n",
       "[3887 rows x 4 columns]"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res_df['abs_diff'] = res_df.difference.abs()\n",
    "\n",
    "res_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Sort the DataFrame\n",
    "\n",
    "Now, let's sort the results by the abs_diff column so that the genes that change the most go to the top. We'll use the [sort_values() function](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.sort_values.html?highlight=sort_values#pandas.DataFrame.sort_values) and pass \"ascending = False\" so that the largest differences show up first.\n",
    "\n",
    "Here, we assign the result to the original DataFrame, replacing it in memory."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>glu2_avg</th>\n",
       "      <th>fru_avg</th>\n",
       "      <th>difference</th>\n",
       "      <th>abs_diff</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>b2169</th>\n",
       "      <td>3.174795</td>\n",
       "      <td>7.211384</td>\n",
       "      <td>4.036589</td>\n",
       "      <td>4.036589</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b2168</th>\n",
       "      <td>3.918281</td>\n",
       "      <td>6.974588</td>\n",
       "      <td>3.056308</td>\n",
       "      <td>3.056308</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b2167</th>\n",
       "      <td>4.494259</td>\n",
       "      <td>7.359883</td>\n",
       "      <td>2.865624</td>\n",
       "      <td>2.865624</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b4245</th>\n",
       "      <td>5.902157</td>\n",
       "      <td>8.319012</td>\n",
       "      <td>2.416855</td>\n",
       "      <td>2.416855</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b4244</th>\n",
       "      <td>5.978385</td>\n",
       "      <td>8.283478</td>\n",
       "      <td>2.305092</td>\n",
       "      <td>2.305092</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b1511</th>\n",
       "      <td>2.790128</td>\n",
       "      <td>2.789732</td>\n",
       "      <td>-0.000396</td>\n",
       "      <td>0.000396</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b4320</th>\n",
       "      <td>3.302079</td>\n",
       "      <td>3.302462</td>\n",
       "      <td>0.000383</td>\n",
       "      <td>0.000383</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b1350</th>\n",
       "      <td>1.686470</td>\n",
       "      <td>1.686152</td>\n",
       "      <td>-0.000318</td>\n",
       "      <td>0.000318</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b3129</th>\n",
       "      <td>4.005529</td>\n",
       "      <td>4.005690</td>\n",
       "      <td>0.000161</td>\n",
       "      <td>0.000161</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b2208</th>\n",
       "      <td>1.623581</td>\n",
       "      <td>1.623594</td>\n",
       "      <td>0.000013</td>\n",
       "      <td>0.000013</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3887 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       glu2_avg   fru_avg  difference  abs_diff\n",
       "b2169  3.174795  7.211384    4.036589  4.036589\n",
       "b2168  3.918281  6.974588    3.056308  3.056308\n",
       "b2167  4.494259  7.359883    2.865624  2.865624\n",
       "b4245  5.902157  8.319012    2.416855  2.416855\n",
       "b4244  5.978385  8.283478    2.305092  2.305092\n",
       "...         ...       ...         ...       ...\n",
       "b1511  2.790128  2.789732   -0.000396  0.000396\n",
       "b4320  3.302079  3.302462    0.000383  0.000383\n",
       "b1350  1.686470  1.686152   -0.000318  0.000318\n",
       "b3129  4.005529  4.005690    0.000161  0.000161\n",
       "b2208  1.623581  1.623594    0.000013  0.000013\n",
       "\n",
       "[3887 rows x 4 columns]"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res_df = res_df.sort_values('abs_diff', ascending = False)\n",
    "\n",
    "res_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The default pandas output above includes the 5 least changed genes, which is pretty useless. Let's use iloc to just grab the top 5 genes instead:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>glu2_avg</th>\n",
       "      <th>fru_avg</th>\n",
       "      <th>difference</th>\n",
       "      <th>abs_diff</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>b2169</th>\n",
       "      <td>3.174795</td>\n",
       "      <td>7.211384</td>\n",
       "      <td>4.036589</td>\n",
       "      <td>4.036589</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b2168</th>\n",
       "      <td>3.918281</td>\n",
       "      <td>6.974588</td>\n",
       "      <td>3.056308</td>\n",
       "      <td>3.056308</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b2167</th>\n",
       "      <td>4.494259</td>\n",
       "      <td>7.359883</td>\n",
       "      <td>2.865624</td>\n",
       "      <td>2.865624</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b4245</th>\n",
       "      <td>5.902157</td>\n",
       "      <td>8.319012</td>\n",
       "      <td>2.416855</td>\n",
       "      <td>2.416855</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b4244</th>\n",
       "      <td>5.978385</td>\n",
       "      <td>8.283478</td>\n",
       "      <td>2.305092</td>\n",
       "      <td>2.305092</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       glu2_avg   fru_avg  difference  abs_diff\n",
       "b2169  3.174795  7.211384    4.036589  4.036589\n",
       "b2168  3.918281  6.974588    3.056308  3.056308\n",
       "b2167  4.494259  7.359883    2.865624  2.865624\n",
       "b4245  5.902157  8.319012    2.416855  2.416855\n",
       "b4244  5.978385  8.283478    2.305092  2.305092"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "top_genes = res_df.iloc[0:5]\n",
    "top_genes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Connecting to Reality: Adding Gene Information\n",
    "\n",
    "We have our top gene results above, so we technically know which genes react most strongly to a switch between glucose and fructose. We have \"information\", but no \"knowledge\". Let's add in some knowledge by finding the gene information relating to these genes:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_product</th>\n",
       "      <th>COG</th>\n",
       "      <th>regulator</th>\n",
       "      <th>operon</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>b2169</th>\n",
       "      <td>fruB</td>\n",
       "      <td>fructose-specific PTS multiphosphoryl transfer...</td>\n",
       "      <td>Carbohydrate transport and metabolism</td>\n",
       "      <td>rpoD,cra</td>\n",
       "      <td>fruBKA</td>\n",
       "      <td>2262365.0</td>\n",
       "      <td>2263495.0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b2168</th>\n",
       "      <td>fruK</td>\n",
       "      <td>1-phosphofructokinase</td>\n",
       "      <td>Nucleotide transport and metabolism</td>\n",
       "      <td>rpoD,cra</td>\n",
       "      <td>fruBKA</td>\n",
       "      <td>2261427.0</td>\n",
       "      <td>2262365.0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b2167</th>\n",
       "      <td>fruA</td>\n",
       "      <td>fructose-specific PTS multiphosphoryl transfer...</td>\n",
       "      <td>Carbohydrate transport and metabolism</td>\n",
       "      <td>rpoD,cra</td>\n",
       "      <td>fruBKA</td>\n",
       "      <td>2259719.0</td>\n",
       "      <td>2261410.0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b4245</th>\n",
       "      <td>pyrB</td>\n",
       "      <td>aspartate carbamoyltransferase catalytic subunit</td>\n",
       "      <td>Nucleotide transport and metabolism</td>\n",
       "      <td>argR,nac,purR</td>\n",
       "      <td>pyrLBI</td>\n",
       "      <td>4471460.0</td>\n",
       "      <td>4472395.0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b4244</th>\n",
       "      <td>pyrI</td>\n",
       "      <td>aspartate carbamoyltransferase%2C PyrI subunit</td>\n",
       "      <td>Nucleotide transport and metabolism</td>\n",
       "      <td>argR,nac,purR</td>\n",
       "      <td>pyrLBI</td>\n",
       "      <td>4470986.0</td>\n",
       "      <td>4471447.0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      gene_name                                       gene_product  \\\n",
       "b2169      fruB  fructose-specific PTS multiphosphoryl transfer...   \n",
       "b2168      fruK                              1-phosphofructokinase   \n",
       "b2167      fruA  fructose-specific PTS multiphosphoryl transfer...   \n",
       "b4245      pyrB   aspartate carbamoyltransferase catalytic subunit   \n",
       "b4244      pyrI     aspartate carbamoyltransferase%2C PyrI subunit   \n",
       "\n",
       "                                         COG      regulator  operon  \\\n",
       "b2169  Carbohydrate transport and metabolism       rpoD,cra  fruBKA   \n",
       "b2168    Nucleotide transport and metabolism       rpoD,cra  fruBKA   \n",
       "b2167  Carbohydrate transport and metabolism       rpoD,cra  fruBKA   \n",
       "b4245    Nucleotide transport and metabolism  argR,nac,purR  pyrLBI   \n",
       "b4244    Nucleotide transport and metabolism  argR,nac,purR  pyrLBI   \n",
       "\n",
       "           start        end strand  \n",
       "b2169  2262365.0  2263495.0      -  \n",
       "b2168  2261427.0  2262365.0      -  \n",
       "b2167  2259719.0  2261410.0      -  \n",
       "b4245  4471460.0  4472395.0      -  \n",
       "b4244  4470986.0  4471447.0      -  "
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "top_genes_info = gene_table.loc[top_genes.index]\n",
    "top_genes_info"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can even combine the data and the gene details by [concatenating](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.concat.html) them together, which can be a useful way to summarize all results in a table:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>glu2_avg</th>\n",
       "      <th>fru_avg</th>\n",
       "      <th>difference</th>\n",
       "      <th>abs_diff</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_product</th>\n",
       "      <th>COG</th>\n",
       "      <th>regulator</th>\n",
       "      <th>operon</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>b2169</th>\n",
       "      <td>3.174795</td>\n",
       "      <td>7.211384</td>\n",
       "      <td>4.036589</td>\n",
       "      <td>4.036589</td>\n",
       "      <td>fruB</td>\n",
       "      <td>fructose-specific PTS multiphosphoryl transfer...</td>\n",
       "      <td>Carbohydrate transport and metabolism</td>\n",
       "      <td>rpoD,cra</td>\n",
       "      <td>fruBKA</td>\n",
       "      <td>2262365.0</td>\n",
       "      <td>2263495.0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b2168</th>\n",
       "      <td>3.918281</td>\n",
       "      <td>6.974588</td>\n",
       "      <td>3.056308</td>\n",
       "      <td>3.056308</td>\n",
       "      <td>fruK</td>\n",
       "      <td>1-phosphofructokinase</td>\n",
       "      <td>Nucleotide transport and metabolism</td>\n",
       "      <td>rpoD,cra</td>\n",
       "      <td>fruBKA</td>\n",
       "      <td>2261427.0</td>\n",
       "      <td>2262365.0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b2167</th>\n",
       "      <td>4.494259</td>\n",
       "      <td>7.359883</td>\n",
       "      <td>2.865624</td>\n",
       "      <td>2.865624</td>\n",
       "      <td>fruA</td>\n",
       "      <td>fructose-specific PTS multiphosphoryl transfer...</td>\n",
       "      <td>Carbohydrate transport and metabolism</td>\n",
       "      <td>rpoD,cra</td>\n",
       "      <td>fruBKA</td>\n",
       "      <td>2259719.0</td>\n",
       "      <td>2261410.0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b4245</th>\n",
       "      <td>5.902157</td>\n",
       "      <td>8.319012</td>\n",
       "      <td>2.416855</td>\n",
       "      <td>2.416855</td>\n",
       "      <td>pyrB</td>\n",
       "      <td>aspartate carbamoyltransferase catalytic subunit</td>\n",
       "      <td>Nucleotide transport and metabolism</td>\n",
       "      <td>argR,nac,purR</td>\n",
       "      <td>pyrLBI</td>\n",
       "      <td>4471460.0</td>\n",
       "      <td>4472395.0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b4244</th>\n",
       "      <td>5.978385</td>\n",
       "      <td>8.283478</td>\n",
       "      <td>2.305092</td>\n",
       "      <td>2.305092</td>\n",
       "      <td>pyrI</td>\n",
       "      <td>aspartate carbamoyltransferase%2C PyrI subunit</td>\n",
       "      <td>Nucleotide transport and metabolism</td>\n",
       "      <td>argR,nac,purR</td>\n",
       "      <td>pyrLBI</td>\n",
       "      <td>4470986.0</td>\n",
       "      <td>4471447.0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       glu2_avg   fru_avg  difference  abs_diff gene_name  \\\n",
       "b2169  3.174795  7.211384    4.036589  4.036589      fruB   \n",
       "b2168  3.918281  6.974588    3.056308  3.056308      fruK   \n",
       "b2167  4.494259  7.359883    2.865624  2.865624      fruA   \n",
       "b4245  5.902157  8.319012    2.416855  2.416855      pyrB   \n",
       "b4244  5.978385  8.283478    2.305092  2.305092      pyrI   \n",
       "\n",
       "                                            gene_product  \\\n",
       "b2169  fructose-specific PTS multiphosphoryl transfer...   \n",
       "b2168                              1-phosphofructokinase   \n",
       "b2167  fructose-specific PTS multiphosphoryl transfer...   \n",
       "b4245   aspartate carbamoyltransferase catalytic subunit   \n",
       "b4244     aspartate carbamoyltransferase%2C PyrI subunit   \n",
       "\n",
       "                                         COG      regulator  operon  \\\n",
       "b2169  Carbohydrate transport and metabolism       rpoD,cra  fruBKA   \n",
       "b2168    Nucleotide transport and metabolism       rpoD,cra  fruBKA   \n",
       "b2167  Carbohydrate transport and metabolism       rpoD,cra  fruBKA   \n",
       "b4245    Nucleotide transport and metabolism  argR,nac,purR  pyrLBI   \n",
       "b4244    Nucleotide transport and metabolism  argR,nac,purR  pyrLBI   \n",
       "\n",
       "           start        end strand  \n",
       "b2169  2262365.0  2263495.0      -  \n",
       "b2168  2261427.0  2262365.0      -  \n",
       "b2167  2259719.0  2261410.0      -  \n",
       "b4245  4471460.0  4472395.0      -  \n",
       "b4244  4470986.0  4471447.0      -  "
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "top_genes = pd.concat([top_genes, top_genes_info], axis = 1)\n",
    "top_genes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "These top genes make sense! Let's discuss why:\n",
    "\n",
    "- the top gene, fruB, is strongly overexpressed in fructose. It is a fructose transporter, assisting in getting fructose into the cell. This would have a clear use in the fructose condition, which makes sense.\n",
    "- the next genes, fruK and fruA, are in the same operon as fruB and are also important for fructose metabolism.\n",
    "- the pyrBI genes are less obvious, but an [EcoCyc search](https://biocyc.org/gene?orgid=ECOLI&id=EG10805) shows that they are involved in the UMP Biosynthesis pathway. You could therefore predict that fructose affects nucleotide synthesis.\n",
    "\n",
    "Connecting some knowledge to your results is a critical step; just finding top genes is useless unless you can gain some insight from them."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": true,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": true,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
